// @spikerko/tools/DynamicBackground.js
import * as THREE from "@jsr/3d__three";
//import * as THREE from "@3d/three";
import { GetShaderUniforms, VertexShader, FragmentShader, DisposeShaderUniforms } from "./DBG_ThreeShaders.js";
// deno-lint-ignore verbatim-module-syntax
import { Maid } from "@jsr/socali__modules/Maid";
import PrefixError from "./PrefixError.js";
const DynamicBackgroundError = new PrefixError({
  name: "DynamicBackgroundError",
  prefix: "DynamicBackground: "
}).Create();
/**
 * DynamicBackground class that implements Giveable interface
 * Creates and manages a THREE.js canvas with animated background
 */ export class DynamicBackground {
  // public properties
  container;
  maid;
  resizeObserver;
  blurAmount;
  transitionDuration;
  rotationSpeed;
  rotationAngle = 0;
  lastFrameTime = 0;
  // Update state
  _isUpdating = false;
  _queuedUpdate = null;
  // Track current values for change detection
  currentImage;
  currentPlaceholderHueShift = 0;
  // THREE.js objects that were previously static
  renderCamera;
  meshGeometry;
  // Cache for blurred cover arts
  blurredCoverArts;
  // deno-lint-ignore no-explicit-any
  plugins;
  clientOptions;
  /**
     * Creates a new DynamicBackground
     * @param options Configuration options
     */ constructor(options = {}){
    this.clientOptions = options;
    // Convert plugins array or object to a normalized object with plugin names as keys
    const pluginsInput = options.plugins ?? [];
    // deno-lint-ignore no-explicit-any
    let pluginsObj = {};
    if (Array.isArray(pluginsInput)) {
      for (const plugin of pluginsInput){
        if (Array.isArray(plugin)) {
          // Tuple format for new plugins: [PluginClass, options]
          const [PluginClass, pluginOptions] = plugin;
          if (PluginClass) {
            const pluginInstance = new PluginClass(pluginOptions);
            if (pluginInstance && typeof pluginInstance === "object" && typeof pluginInstance.name === "string") {
              pluginsObj[pluginInstance.name] = pluginInstance;
            }
          }
        } else {
          // Ensure plugin is an object and has a string 'name' property
          if (plugin && typeof plugin === "object" && typeof plugin.name === "string") {
            pluginsObj[plugin.name] = plugin;
          }
        }
      }
    } else if (typeof pluginsInput === "object" && pluginsInput !== null && !Array.isArray(pluginsInput)) {
      // If already an object, shallow copy (only if it's a plain object)
      pluginsObj = Object.assign({}, pluginsInput);
    }
    this.plugins = pluginsObj;
    // Set default values
    this.blurAmount = options.blur ?? 40;
    this.rotationSpeed = options.speed ?? 0.2;
    this.blurredCoverArts = options.coverArtCache ?? new Map();
    // Handle transition option (can be boolean or number)
    if (typeof options.transition === 'boolean') {
      this.transitionDuration = options.transition ? 0.2 : 0;
    } else {
      this.transitionDuration = options.transition ?? 0.2;
    }
    // Create or use provided maid
    this.maid = options.maid ?? new Maid();
    // Initialize THREE.js objects
    this.initThreeObjects();
    // Register THREE.js geometry with Maid
    this.maid.Give(()=>{
      if (this.meshGeometry) {
        this.meshGeometry.dispose();
      }
    });
    // Create the renderer
    const renderer = new THREE.WebGLRenderer({
      alpha: true,
      antialias: true,
      powerPreference: 'default',
      preserveDrawingBuffer: false
    });
    // Setup container
    this.container = renderer.domElement;
    // Create scene and materials
    const renderScene = new THREE.Scene();
    const materialUniforms = GetShaderUniforms();
    const meshMaterial = new THREE.ShaderMaterial({
      uniforms: materialUniforms,
      vertexShader: VertexShader,
      fragmentShader: FragmentShader
    });
    this.container.material = meshMaterial;
    // Create mesh and add to scene
    const sceneMesh = new THREE.Mesh(this.meshGeometry, meshMaterial);
    renderScene.add(sceneMesh);
    // Set container properties
    this.container.renderer = renderer;
    this.container.scene = renderScene;
    this.container.uniforms = materialUniforms;
    // Set initial rotation speed
    this.container.uniforms.RotationSpeed.value = this.rotationSpeed;
    // Register renderer cleanup with Maid
    this.maid.Give(()=>{
      if (this.container.renderer) {
        this.container.renderer.dispose();
        const gl = this.container.renderer.getContext();
        if (gl && !gl.isContextLost()) {
          const loseContext = gl.getExtension('WEBGL_lose_context');
          if (loseContext) loseContext.loseContext();
        }
        this.container.renderer = undefined;
      }
    });
    // Register material cleanup with Maid
    this.maid.Give(()=>{
      if (this.container.material) {
        this.container.material.dispose();
        this.container.material = undefined;
      }
    });
    // Register texture cleanup with Maid
    this.maid.Give(()=>{
      if (this.container.texture) {
        this.container.texture.dispose();
        this.container.texture = undefined;
      }
    });
    // Register shader uniforms cleanup with Maid
    this.maid.Give(()=>{
      if (this.container.uniforms) {
        DisposeShaderUniforms(this.container.uniforms);
      }
    });
    // Register animation frame cleanup with Maid
    this.maid.Give(()=>{
      if (this.container.animationFrameId) {
        cancelAnimationFrame(this.container.animationFrameId);
        this.container.animationFrameId = undefined;
      }
    });
    // Register DOM element removal with Maid
    this.maid.Give(()=>{
      if (this.container.parentElement) {
        this.container.remove();
      }
    });
    // Register blurred cover arts cache cleanup with Maid
    this.maid.Give(()=>{
      if (!options.coverArtCache) {
        this.blurredCoverArts.clear();
      }
    });
    // Still keep the comprehensive cleanup as a fallback
    this.maid.Give(()=>this.cleanup());
    Object.values(this.plugins).map(async (plugin)=>{
      if (!plugin) return;
      this.maid.Give(()=>plugin.Destroy());
      await plugin.initialize({
        ClientOptions: options,
        InternalContent: this
      });
    });
  }
  /**
     * Initialize Three.js static objects
     * This is now an instance method that creates objects for this instance only
     */ initThreeObjects() {
    this.renderCamera = new THREE.OrthographicCamera(-1, 1, 1, -1, 0.1, 10);
    this.renderCamera.position.z = 1;
    this.meshGeometry = new THREE.PlaneGeometry(2, 2);
  }
  /**
     * Updates the background with a new image
     * @param options Update options
     * @returns Promise that resolves when the update is complete
     */ async Update(options) {
    this._queuedUpdate = options;
    if (this._isUpdating) {
      return;
    }
    this._isUpdating = true;
    try {
      while(this._queuedUpdate){
        const currentUpdateOptions = this._queuedUpdate;
        this._queuedUpdate = null;
        await this._performUpdate(currentUpdateOptions);
      }
    } finally{
      this._isUpdating = false;
    }
  }
  async _performUpdate(options) {
    // Don't update if maid is destroyed
    if (this.maid.IsDestroyed()) return;
    const { image: newImage, placeholderHueShift = 0, blur = this.blurAmount, speed = this.rotationSpeed } = options;
    const image = newImage ?? this.currentImage;
    if (!image || typeof image !== 'string') {
      throw new DynamicBackgroundError("Image must be a string and is required for the first update.");
    }
    // Check if anything has changed
    const imageChanged = image !== this.currentImage;
    const hueShiftChanged = placeholderHueShift !== this.currentPlaceholderHueShift;
    const blurChanged = blur !== this.blurAmount;
    const oldSpeed = this.rotationSpeed;
    const speedChanged = speed !== oldSpeed;
    // If nothing has changed, return early
    if (!imageChanged && !hueShiftChanged && !blurChanged && !speedChanged) {
      return;
    }
    if (blurChanged || hueShiftChanged) {
      this.blurredCoverArts.delete(image);
    }
    // Update stored values
    this.blurAmount = blur;
    this.rotationSpeed = speed;
    this.currentImage = image;
    this.currentPlaceholderHueShift = placeholderHueShift;
    // If no current texture, initialize it
    if (!this.container.texture) {
      await this.initializeTexture(image, placeholderHueShift);
      return;
    }
    // Cancel any existing animation frame
    if (this.container.animationFrameId) {
      cancelAnimationFrame(this.container.animationFrameId);
      this.container.animationFrameId = undefined;
    }
    // Get the new blurred cover art
    const newBlurredCover = await this.getBlurredCoverArt(image, placeholderHueShift);
    // Create a new texture
    const newTexture = new THREE.CanvasTexture(newBlurredCover);
    newTexture.minFilter = THREE.NearestFilter;
    newTexture.magFilter = THREE.NearestFilter;
    newTexture.needsUpdate = true;
    // Dispose of any existing new texture
    if (this.container.uniforms.NewBlurredCoverArt.value) {
      this.container.uniforms.NewBlurredCoverArt.value.dispose();
    }
    // Set the new texture
    this.container.uniforms.NewBlurredCoverArt.value = newTexture;
    this.container.uniforms.RotationSpeed.value = oldSpeed;
    // Force a render to ensure the texture is loaded before starting the animation
    if (this.container.renderer && this.container.scene) {
      this.container.renderer.render(this.container.scene, this.renderCamera);
    }
    // Reset transition progress
    this.container.uniforms.TransitionProgress.value = 0;
    // Skip animation if transition duration is 0
    if (this.transitionDuration <= 0) {
      this.completeTransition(newTexture, image);
      return;
    }
    // Animate the transition
    await this.animateTransition(newTexture, image, oldSpeed);
  }
  /**
     * Returns the canvas element for external use
     * @returns HTMLElement containing the THREE.js canvas
     */ GetCanvasElement() {
    return this.container;
  }
  /**
     * Destroys the background and cleans up all resources
     */ Destroy() {
    // Check if maid is already destroyed
    if (this.maid.IsDestroyed()) {
      this.cleanup();
      return;
    }
    // Run cleanup regardless of maid ownership
    // This ensures all resources are properly cleaned up
    this.cleanup();
    // Always destroy the maid, regardless of ownership
    // This is important for proper resource cleanup
    this.maid.Destroy();
  }
  /**
     * Initializes the texture for the first time
     * @param imageCoverUrl URL of the image to use
     * @param placeholderHueShift Optional hue shift for placeholder images
     */ async initializeTexture(imageCoverUrl, placeholderHueShift = 0) {
    const blurredCover = await this.getBlurredCoverArt(imageCoverUrl, placeholderHueShift);
    const texture = new THREE.CanvasTexture(blurredCover);
    texture.minFilter = THREE.NearestFilter;
    texture.magFilter = THREE.NearestFilter;
    this.container.texture = texture;
    this.container.uniforms.BlurredCoverArt.value = texture;
    this.container.uniforms.Time.value = 0;
    this.container.uniforms.RotationSpeed.value = this.rotationSpeed;
    this.container.setAttribute("data-cover-id", imageCoverUrl);
    // Store current values for change detection
    this.currentImage = imageCoverUrl;
    this.currentPlaceholderHueShift = placeholderHueShift;
  }
  /**
     * Animates the transition between textures
     * @param newTexture The new texture to transition to
     * @param newCoverArtUrl URL of the new cover art
     */ animateTransition(newTexture, newCoverArtUrl, oldSpeed) {
    return new Promise((resolve)=>{
      const newSpeed = this.rotationSpeed;
      const duration = this.transitionDuration * 1000;
      let startTime = null;
      const animationState = {
        canceled: false,
        frameId: 0
      };
      let cleanupKey;
      if (!this.maid.IsDestroyed()) {
        cleanupKey = this.maid.Give(()=>{
          animationState.canceled = true;
          cancelAnimationFrame(animationState.frameId);
          if (this.container.uniforms) {
            this.container.uniforms.TransitionProgress.value = 0;
            this.container.uniforms.RotationSpeed.value = this.rotationSpeed;
          }
        });
      }
      this.lastFrameTime = performance.now();
      const animate = (timestamp)=>{
        if (animationState.canceled) {
          resolve();
          return;
        }
        if (startTime === null) {
          startTime = timestamp;
        }
        const elapsedTime = timestamp - startTime;
        const progress = Math.min(elapsedTime / duration, 1);
        const deltaTime = (timestamp - this.lastFrameTime) / 1000;
        this.lastFrameTime = timestamp;
        const animatedSpeed = oldSpeed + (newSpeed - oldSpeed) * progress;
        this.rotationAngle += deltaTime * animatedSpeed;
        this.container.uniforms.RotationAngle.value = this.rotationAngle;
        this.container.uniforms.TransitionProgress.value = progress;
        if (this.container.renderer && this.container.scene) {
          this.container.renderer.render(this.container.scene, this.renderCamera);
        }
        if (progress < 1) {
          animationState.frameId = requestAnimationFrame(animate);
        } else {
          if (animationState.canceled) {
            resolve();
            return;
          }
          this.completeTransition(newTexture, newCoverArtUrl);
          if (cleanupKey !== undefined && !this.maid.IsDestroyed()) {
            this.maid.Clean(cleanupKey);
          }
          resolve();
        }
      };
      animationState.frameId = requestAnimationFrame(animate);
    });
  }
  /**
     * Completes the transition by swapping textures
     * @param newTexture The new texture to use
     * @param newCoverArtUrl URL of the new cover art
     */ completeTransition(newTexture, newCoverArtUrl) {
    // When animation is complete, swap textures
    if (this.container.texture) {
      this.container.texture.dispose();
    }
    this.container.texture = newTexture;
    this.container.uniforms.BlurredCoverArt.value = newTexture;
    this.container.uniforms.NewBlurredCoverArt.value = null;
    this.container.uniforms.TransitionProgress.value = 0;
    this.container.uniforms.RotationSpeed.value = this.rotationSpeed;
    this.container.setAttribute("data-cover-id", newCoverArtUrl);
    // Update current image for change detection
    this.currentImage = newCoverArtUrl;
    // Force a render to ensure the new texture is displayed
    if (this.container.renderer && this.container.scene) {
      this.container.renderer.render(this.container.scene, this.renderCamera);
    }
    // Start animation loop
    this.startAnimation();
  }
  /**
     * Starts the animation loop
     */ startAnimation() {
    // Cancel any existing animation
    if (this.container.animationFrameId) {
      cancelAnimationFrame(this.container.animationFrameId);
      this.container.animationFrameId = undefined;
    }
    this.lastFrameTime = performance.now();
    const animate = (time)=>{
      // Check if container and renderer still exist
      if (!this.container || !this.container.renderer || this.container.renderer.getContext()?.isContextLost()) {
        if (this.container?.animationFrameId) {
          cancelAnimationFrame(this.container.animationFrameId);
          this.container.animationFrameId = undefined;
        }
        return;
      }
      // Check if renderCamera exists (it might have been cleaned up)
      if (!this.renderCamera) return;
      const deltaTime = (time - this.lastFrameTime) / 1000;
      this.lastFrameTime = time;
      this.rotationAngle += deltaTime * this.rotationSpeed;
      this.container.uniforms.RotationAngle.value = this.rotationAngle;
      this.container.renderer.render(this.container.scene, this.renderCamera);
      this.container.animationFrameId = requestAnimationFrame(animate);
    };
    animate(performance.now());
  }
  /**
     * Updates the container dimensions when parent element size changes
     * @param width New width
     * @param height New height
     */ updateContainerDimensions(width, height) {
    const { renderer, scene, uniforms } = this.container;
    renderer.setSize(width, height);
    renderer.setPixelRatio(globalThis.devicePixelRatio);
    const scaledWidth = width * globalThis.devicePixelRatio;
    const scaledHeight = height * globalThis.devicePixelRatio;
    const largestAxis = scaledWidth > scaledHeight ? "X" : "Y";
    const largestAxisSize = scaledWidth > scaledHeight ? scaledWidth : scaledHeight;
    uniforms.BackgroundCircleOrigin.value.set(scaledWidth / 2, scaledHeight / 2);
    uniforms.BackgroundCircleRadius.value = largestAxisSize * 1.5;
    uniforms.CenterCircleOrigin.value.set(scaledWidth / 2, scaledHeight / 2);
    uniforms.CenterCircleRadius.value = largestAxisSize * (largestAxis === "X" ? 1 : 0.75);
    uniforms.LeftCircleOrigin.value.set(0, scaledHeight);
    uniforms.LeftCircleRadius.value = largestAxisSize * 0.75;
    uniforms.RightCircleOrigin.value.set(scaledWidth, 0);
    uniforms.RightCircleRadius.value = largestAxisSize * (largestAxis === "X" ? 0.65 : 0.5);
    renderer.render(scene, this.renderCamera);
    this.startAnimation();
  }
  /**
     * Appends the background to a parent element
     * @param element Parent element to append the background to
     */ AppendToElement(element) {
    // Don't append if maid is destroyed
    if (this.maid.IsDestroyed()) return;
    // Remove from current parent if any
    if (this.container.parentElement) {
      this.container.remove();
    }
    // Disconnect existing resize observer
    if (this.resizeObserver) {
      this.resizeObserver.disconnect();
      this.resizeObserver = undefined;
    }
    // Add to new parent
    element.appendChild(this.container);
    // Create new resize observer
    this.resizeObserver = new ResizeObserver((entries)=>{
      for (const entry of entries){
        const width = Math.max(entry.contentRect.width, 500);
        const height = Math.max(entry.contentRect.height, 500);
        this.updateContainerDimensions(width, height);
      }
    });
    // Add to maid for proper cleanup if maid is not destroyed
    if (!this.maid.IsDestroyed()) {
      this.maid.Give(()=>{
        if (this.resizeObserver) {
          this.resizeObserver.disconnect();
          this.resizeObserver = undefined;
        }
      });
    }
    // Start observing
    this.resizeObserver.observe(element);
    // Initial size update
    const width = Math.max(element.clientWidth, 500);
    const height = Math.max(element.clientHeight, 500);
    this.updateContainerDimensions(width, height);
  }
  /**
     * Gets a blurred cover art from the URL
     * @param coverArtUrl URL of the cover art
     * @param placeholderHueShift Optional hue shift for placeholder images
     * @returns Promise that resolves to an OffscreenCanvas with the blurred image
     */ async getBlurredCoverArt(coverArtUrl, placeholderHueShift = 0) {
    if (this.blurredCoverArts.has(coverArtUrl)) {
      // Refresh LRU order: delete and re-set the key to move it to the end
      const canvas = this.blurredCoverArts.get(coverArtUrl);
      // Only refresh if a cacheLimit is defined and reached
      if (typeof this.clientOptions?.cacheLimit === "number" && this.blurredCoverArts.size >= this.clientOptions.cacheLimit) {
        this.blurredCoverArts.delete(coverArtUrl);
        this.blurredCoverArts.set(coverArtUrl, canvas);
      }
      return canvas;
    }
    const image = new Image();
    image.src = coverArtUrl;
    if (coverArtUrl.includes("https://") || coverArtUrl.includes("http://")) {
      image.crossOrigin = "anonymous";
    }
    await image.decode();
    const originalSize = Math.min(image.width, image.height);
    const resizedBlurAmount = this.blurAmount * (originalSize / 640);
    const blurExtent = Math.ceil(3 * resizedBlurAmount);
    const circleCanvas = new OffscreenCanvas(originalSize, originalSize);
    const circleCtx = circleCanvas.getContext('2d');
    circleCtx.beginPath();
    circleCtx.arc(originalSize / 2, originalSize / 2, originalSize / 2, 0, Math.PI * 2);
    circleCtx.closePath();
    circleCtx.clip();
    circleCtx.drawImage(image, (image.width - originalSize) / 2, (image.height - originalSize) / 2, originalSize, originalSize, 0, 0, originalSize, originalSize);
    const padding = blurExtent * 1.5;
    const expandedSize = originalSize + padding;
    const blurredCanvas = new OffscreenCanvas(expandedSize, expandedSize);
    const blurredCtx = blurredCanvas.getContext('2d');
    blurredCtx.filter = `blur(${resizedBlurAmount}px) hue-rotate(${placeholderHueShift}deg)`;
    blurredCtx.drawImage(circleCanvas, padding / 2, padding / 2);
    // Maintain a cache limit of 10 entries (LRU)
    if (typeof this.clientOptions?.cacheLimit === "number" && this.blurredCoverArts.size >= this.clientOptions.cacheLimit) {
      // Delete the oldest entry.
      const oldestKey = this.blurredCoverArts.keys().next().value;
      if (oldestKey !== undefined) {
        this.blurredCoverArts.delete(oldestKey);
      }
    }
    this.blurredCoverArts.set(coverArtUrl, blurredCanvas);
    return blurredCanvas;
  }
  /**
     * Prefetches an image to blurred cover arts (makes it seem faster on image change - since this is for prefetching)
     * @param imageUrl URL of the image
     * @param placeholderHueShift Optional hue shift for placeholder images
     * @returns Promise that resolves to an OffscreenCanvas with the blurred image
     */ async PrefetchImage(imageUrl, placeholderHueShift = 0) {
    const blurredCoverArt = await this.getBlurredCoverArt(imageUrl, placeholderHueShift);
    return blurredCoverArt;
  }
  /**
     * Cleans up all resources used by the background
     * This is a fallback cleanup method in case individual Maid cleanups fail
     */ cleanup() {
    // Disconnect resize observer
    if (this.resizeObserver) {
      this.resizeObserver.disconnect();
      this.resizeObserver = undefined;
    }
    // Cancel animation frame
    if (this.container.animationFrameId) {
      cancelAnimationFrame(this.container.animationFrameId);
      this.container.animationFrameId = undefined;
    }
    // Dispose of shader uniforms
    if (this.container.uniforms) {
      DisposeShaderUniforms(this.container.uniforms);
    }
    // Dispose of material
    if (this.container.material) {
      this.container.material.dispose();
      this.container.material = undefined;
    }
    // Dispose of texture
    if (this.container.texture) {
      this.container.texture.dispose();
      this.container.texture = undefined;
    }
    // Clean up scene
    if (this.container.scene) {
      // Dispose all objects in the scene
      this.container.scene.traverse((object)=>{
        if (object instanceof THREE.Mesh) {
          if (object.geometry) object.geometry.dispose();
          if (object.material) {
            if (Array.isArray(object.material)) {
              object.material.forEach((material)=>material.dispose());
            } else {
              object.material.dispose();
            }
          }
        }
      });
    }
    // Dispose of renderer
    if (this.container.renderer) {
      this.container.renderer.dispose();
      const gl = this.container.renderer.getContext();
      if (gl && !gl.isContextLost()) {
        const loseContext = gl.getExtension('WEBGL_lose_context');
        if (loseContext) loseContext.loseContext();
      }
      this.container.renderer = undefined;
    }
    // Remove from parent
    if (this.container) {
      this.container.remove();
    }
    // Clean up THREE.js objects
    if (this.meshGeometry) {
      this.meshGeometry.dispose();
    }
    // Reset tracking variables
    this.currentImage = undefined;
    this.currentPlaceholderHueShift = 0;
  }
}
