// @spikerko/tools/TempoPlugin.js
import { Maid } from "@jsr/socali__modules/Maid";
import { GetExpireStore } from "./Cache.js";
const AudioDataStore = GetExpireStore("SpikerkoTools_TempoPlugin_AudioData", 1, {
  Unit: "Weeks",
  Duration: 2
});
export class TempoPlugin {
  name = "TempoPlugin";
  maid = new Maid();
  SongChangeSignal;
  getSongId;
  getSongPosition;
  getAccessToken;
  dynamicBg;
  initialized = false;
  // deno-lint-ignore no-explicit-any
  audioDataCache = new Map();
  speedAnimation;
  speedAnimationFunction;
  processedSections;
  lastPaused = false;
  lastSpeed = 0;
  getPaused;
  clientOptions;
  audioDataAbortController;
  options;
  // Internally memoized min/max
  minSpeed = 0.2;
  maxSpeed = 0.525;
  constructor(options){
    this.maid.Give(()=>{
      this.audioDataCache.clear();
      this.audioDataAbortController?.abort();
    });
    this.options = options;
    this.SongChangeSignal = this.options.SongChangeSignal;
    this.getSongId = this.options.getSongId;
    this.getSongPosition = this.options.getSongPosition;
    this.getAccessToken = this.options.getAccessToken;
    this.getPaused = this.options.getPaused;
    this.animationLoop = this.animationLoop.bind(this);
    // Get custom min/max or default
    if (typeof options.minSpeed === "number") {
      this.minSpeed = options.minSpeed;
    }
    if (typeof options.maxSpeed === "number") {
      this.maxSpeed = options.maxSpeed;
    }
  }
  initialize(options) {
    if (this.initialized) throw new Error("TempoPlugin was already initialized");
    this.dynamicBg = options.InternalContent;
    this.clientOptions = options.ClientOptions;
    this.lastSpeed = this.dynamicBg.rotationSpeed ?? 0;
    this.initialized = true;
    const initiateProcess = async ()=>{
      try {
        this.processedSections = undefined;
        await this.processSections();
        this.animate();
      } catch (error) {
        console.error("TempoPlugin: Failed to process song sections. Animation not started.", error);
        if (this.speedAnimation) {
          cancelAnimationFrame(this.speedAnimation);
          this.speedAnimation = undefined;
        }
        this.processedSections = undefined;
      }
    };
    setTimeout(()=>{
      initiateProcess();
    }, 50);
    const conn = this.SongChangeSignal?.Connect(initiateProcess);
    this.maid.Give(()=>conn?.Disconnect());
  }
  isInitialized() {
    return this.initialized;
  }
  async getAudioData() {
    if (!this.initialized) throw new Error("TempoPlugin hasn't been initialized yet");
    if (!this.getSongId) throw new Error("TempoPlugin: getSongId() is undefined");
    const songId = this.getSongId();
    this.audioDataAbortController?.abort();
    this.audioDataAbortController = new AbortController();
    const isMapCached = this.audioDataCache.has(songId);
    if (isMapCached) {
      return this.audioDataCache.get(songId);
    }
    const cached = await AudioDataStore.GetItem(songId);
    if (cached) {
      return cached;
    }
    const signal = this.audioDataAbortController.signal;
    if (!this.getAccessToken) throw new Error("TempoPlugin: getAccessToken() is undefined");
    const accessToken = await this.getAccessToken();
    if (!accessToken) throw new Error("TempoPlugin: Access Token missing");
    try {
      const req = await fetch(`https://api.spotify.com/v1/audio-analysis/${songId}`, {
        method: "GET",
        headers: {
          Authorization: accessToken
        },
        signal
      });
      if (req.status !== 200) {
        throw new Error("TempoPlugin: Fetch request failed");
      }
      const res = await req.json();
      if (!res) throw new Error("TempoPlugin: Fetch request failed - content missing");
      this.audioDataCache.set(songId, res);
      await AudioDataStore.SetItem(songId, res);
      this.audioDataAbortController = undefined;
      return res;
    // deno-lint-ignore no-explicit-any
    } catch (err) {
      if (err.name === 'AbortError') {
        console.warn('TempoPlugin: Previous getAudioData request aborted');
        return;
      }
      this.audioDataAbortController = undefined;
      throw new Error("TempoPlugin: Getting Audio Data failed");
    }
  }
  async processSections() {
    const audioData = await this.getAudioData();
    if (!audioData) {
      this.processedSections = undefined;
      return;
    }
    // deno-lint-ignore no-explicit-any
    const tempos = audioData.sections.map((s)=>s.tempo);
    const minTempo = Math.min(...tempos);
    const maxTempo = Math.max(...tempos);
    const tempoRange = maxTempo - minTempo;
    // Use minSpeed/maxSpeed from options if provided
    const newMinSpeed = this.minSpeed;
    const newMaxSpeed = this.maxSpeed;
    // deno-lint-ignore no-explicit-any
    this.processedSections = audioData?.sections?.map((section, index, arr)=>{
      let speed;
      if (tempoRange === 0) {
        speed = newMinSpeed;
      } else {
        speed = newMinSpeed + (newMaxSpeed - newMinSpeed) * (section.tempo - minTempo) / tempoRange;
      }
      return {
        start: section.start,
        end: arr?.[index + 1]?.start ?? section.start + section.duration,
        duration: section.duration,
        tempo: section.tempo,
        speed
      };
    });
  }
  async animationLoop() {
    if (!this.getSongPosition) {
      console.error("TempoPlugin: getSongPosition() is undefined");
      throw new Error("TempoPlugin: getSongPosition() is undefined");
    }
    if (!this.processedSections) {
      //console.warn("TempoPlugin: this.processedSections is undefined, skipping frame.");
      this.speedAnimation = requestAnimationFrame(this.animationLoop);
      return;
    }
    const audioPosition = this.getSongPosition();
    for (const section of this.processedSections){
      const start = section.start;
      const end = section.end;
      const isActive = audioPosition >= start && audioPosition < end;
      if (isActive) {
        if (!this.getPaused) {
          console.error("TempoPlugin: getPaused() is undefined");
          throw new Error("TempoPlugin: getPaused() is undefined");
        }
        if (!this.dynamicBg) {
          console.error("TempoPlugin: dynamicBg() is undefined");
          throw new Error("TempoPlugin: dynamicBg() is undefined");
        }
        if (this.lastPaused !== this.getPaused()) {
          await this.dynamicBg.Update({
            speed: 0
          });
          this.lastSpeed = -1;
        }
        this.lastPaused = this.getPaused();
        if (this.lastSpeed !== section.speed && !this.lastPaused) {
          await this.dynamicBg.Update({
            speed: section.speed
          });
        }
        this.lastSpeed = section.speed;
        break;
      }
    }
    this.speedAnimation = requestAnimationFrame(this.animationLoop);
  }
  animate() {
    if (this.speedAnimation) {
      // console.error("TempoPlugin: Speed animation is already running");
      return;
    }
    this.speedAnimation = requestAnimationFrame(this.animationLoop);
  }
  Destroy() {
    this.initialized = false;
    this.lastSpeed = -1;
    this.lastPaused = false;
    this.processedSections = undefined;
    this.audioDataAbortController?.abort();
    this.audioDataAbortController = undefined;
    if (this.speedAnimation) {
      cancelAnimationFrame(this.speedAnimation);
      this.speedAnimation = undefined;
    }
    this.maid.Destroy();
  }
}
export default function(options) {
  return [
    TempoPlugin,
    options
  ];
}
