// @spikerko/web-modules/Animator.js
import { Maid } from "./Maid.js";
export default class Animator {
  from;
  to;
  duration;
  currentProgress;
  startTime = null;
  pausedTime = null;
  animationFrameId = null;
  events = {};
  isDestroyed = false;
  maid;
  reversed = false;
  constructor(from, to, duration){
    this.from = from;
    this.to = to;
    this.duration = duration * 1000; // Convert to milliseconds
    this.currentProgress = from;
    this.maid = new Maid();
  }
  emit(event, progress) {
    if (this.events[event] && !this.isDestroyed) {
      const callback = this.events[event];
      callback?.(progress ?? this.currentProgress, this.from, this.to);
    }
  }
  on(event, callback) {
    this.events[event] = callback;
  }
  Start() {
    if (this.isDestroyed) return;
    this.startTime = performance.now();
    this.animate();
  }
  animate() {
    if (this.isDestroyed || this.startTime === null) return;
    const now = performance.now();
    const elapsed = now - this.startTime;
    const t = Math.min(elapsed / this.duration, 1);
    const startValue = this.reversed ? this.to : this.from;
    const endValue = this.reversed ? this.from : this.to;
    this.currentProgress = startValue + (endValue - startValue) * t;
    this.emit("progress", this.currentProgress);
    if (t < 1) {
      this.animationFrameId = requestAnimationFrame(()=>this.animate());
      this.maid.Give(()=>cancelAnimationFrame(this.animationFrameId));
    } else {
      this.emit("finish");
      this.reset();
    }
  }
  Pause() {
    if (this.isDestroyed || this.animationFrameId === null) return;
    cancelAnimationFrame(this.animationFrameId);
    this.pausedTime = performance.now();
    this.emit("pause", this.currentProgress);
  }
  Resume() {
    if (this.isDestroyed || this.pausedTime === null) return;
    const pausedDuration = performance.now() - this.pausedTime;
    if (this.startTime !== null) this.startTime += pausedDuration;
    this.pausedTime = null;
    this.emit("resume", this.currentProgress);
    this.animate();
  }
  Restart() {
    if (this.isDestroyed) return;
    this.reset();
    this.emit("restart", this.currentProgress);
    this.Start();
  }
  Reverse() {
    if (this.isDestroyed) return;
    this.reversed = !this.reversed;
    this.emit("reverse", this.currentProgress);
  }
  Destroy() {
    if (this.isDestroyed) return;
    this.emit("destroy");
    this.maid.Destroy();
    this.reset();
    this.isDestroyed = true;
  }
  reset() {
    this.startTime = null;
    this.pausedTime = null;
    this.animationFrameId = null;
  }
}
