// @spikerko/web-modules/FreeArray.js
// Packages
import { GetUniqueId } from "./UniqueId.js";
// Class
class FreeArray {
  // Private Properties
  Items;
  DestroyedState;
  // Constructor
  constructor(){
    // Create our list of items
    this.Items = new Map();
    // Store our initial destroyed state
    this.DestroyedState = false;
  }
  // Public Methods
  Push(item) {
    // Generate our key
    const key = GetUniqueId();
    // Store our item
    this.Items.set(key, item);
    // Return our key
    return key;
  }
  Get(key) {
    return this.Items.get(key);
  }
  Remove(key) {
    // Make sure we have an item
    const item = this.Items.get(key);
    if (item !== undefined) {
      // Remove the item
      this.Items.delete(key);
      // Return our item
      return item;
    }
  }
  GetIterator() {
    return this.Items.entries();
  }
  IsDestroyed() {
    return this.DestroyedState;
  }
  // Deconstructor
  Destroy() {
    // Make sure we aren't already destroyed
    if (this.DestroyedState) {
      return;
    }
    // Mark that we are destroyed
    this.DestroyedState = true;
  }
}
// Export
export { FreeArray };
