// @spikerko/web-modules/Maid.js
// Packages
import { GetUniqueId } from "./UniqueId.js";
import { Signal, IsConnection } from "./Signal.js";
import { IsScheduled, Cancel } from "./Scheduler.js";
class Giveable {
}
// Helper Methods
const IsGiveable = (item)=>{
  return "Destroy" in item;
};
// Class
class Maid {
  // Private Properties
  Items;
  DestroyedState;
  // Signals
  DestroyingSignal;
  CleanedSignal;
  DestroyedSignal;
  // Events
  Destroying;
  Cleaned;
  Destroyed;
  // Constructor
  constructor(){
    // Create our list of items
    this.Items = new Map();
    // Store our initial destroyed state
    this.DestroyedState = false;
    // Create our signals/events
    {
      // Signals
      this.DestroyingSignal = new Signal();
      this.CleanedSignal = new Signal();
      this.DestroyedSignal = new Signal();
      // Events
      this.Destroying = this.DestroyingSignal.GetEvent();
      this.Cleaned = this.CleanedSignal.GetEvent();
      this.Destroyed = this.DestroyedSignal.GetEvent();
    }
  }
  // Private Methods
  CleanItem(item) {
    // Check if we're a maid
    if (IsGiveable(item)) {
      item.Destroy();
    } else if (IsScheduled(item)) {
      Cancel(item);
    } else if (item instanceof MutationObserver || item instanceof ResizeObserver) {
      item.disconnect();
    } else if (IsConnection(item)) {
      item.Disconnect();
    } else if (item instanceof Element) {
      item.remove();
    } else if (typeof item === "function") {
      item();
    } else {
      console.warn("UNSUPPORTED MAID ITEM", typeof item, item);
    }
  }
  // Public Methods
  Give(item, key) {
    // If we're already destroyed then we can just clean the item immediately
    if (this.DestroyedState) {
      this.CleanItem(item);
      return item;
    }
    // Determine our final-key
    const finalKey = key ?? GetUniqueId();
    // Check if we already exist
    if (this.Has(finalKey)) {
      // Clean our previous item
      this.Clean(finalKey);
    }
    // Now store ourselves
    this.Items.set(finalKey, item);
    // Return our item for ease-of-use
    return item;
  }
  GiveItems(...args) {
    // Loop through all of our items
    for (const item of args){
      // Give the item
      this.Give(item);
    }
    // Return back our items
    return args;
  }
  Get(key) {
    return this.DestroyedState ? undefined : this.Items.get(key);
  }
  Has(key) {
    return this.DestroyedState ? false : this.Items.has(key);
  }
  Clean(key) {
    // If we are destroyed then we are already cleaned up everything
    if (this.DestroyedState) {
      return;
    }
    // First determine if we have the item
    const item = this.Items.get(key);
    if (item !== undefined) {
      // Remove the key
      this.Items.delete(key);
      // Clean the item
      this.CleanItem(item);
    }
  }
  CleanUp() {
    // If we're already destroyed then we're already cleaned-up
    if (this.DestroyedState) {
      return;
    }
    // Loop through all of our items
    for (const [key, _] of this.Items){
      // Clean the item
      this.Clean(key);
    }
    // Make sure we aren't destroyed prior to firing
    if (this.DestroyedState === false) {
      this.CleanedSignal.Fire();
    }
  }
  IsDestroyed() {
    return this.DestroyedState;
  }
  // Deconstructor
  Destroy() {
    // Make sure we don't perform twice
    if (this.DestroyedState === false) {
      // Fire our Destroying signal
      this.DestroyingSignal.Fire();
      // Clean out all our items
      this.CleanUp();
      // Set our destroyed state
      this.DestroyedState = true;
      // Fire our destroyed signal
      this.DestroyedSignal.Fire();
      // Now destroy all our signals
      this.DestroyingSignal.Destroy();
      this.CleanedSignal.Destroy();
      this.DestroyedSignal.Destroy();
    }
  }
}
// Export our maid class
export { Maid, Giveable };
