// @spikerko/web-modules/Scheduler.js
// The following image is a visual representation of the stages in a single frame
// https://miro.medium.com/v2/resize:fit:1400/format:webp/1*ad-k5hYKQnRQJF8tv8BIqg.png
// Define a simple and easy data format for keeping track of scheduled tasks
export const Cancel = (scheduled)=>{
  if (scheduled[2]) {
    return; // Prevents any logic processing on a cancelled task
  }
  scheduled[2] = true;
  switch(scheduled[0]){
    case 0:
      globalThis.clearTimeout(scheduled[1]);
      break;
    case 1:
      globalThis.clearInterval(scheduled[1]);
      break;
    case 2:
      globalThis.cancelAnimationFrame(scheduled[1]);
      break;
  }
};
// Handle Javascript Timer Scheduling
export const Timeout = (seconds, callback)=>{
  return [
    0,
    setTimeout(callback, seconds * 1000)
  ];
};
export const Interval = (everySeconds, callback)=>{
  return [
    1,
    setInterval(callback, everySeconds * 1000)
  ];
};
// Now handle Browser Event Scheduling
export const OnPreRender = (callback)=>{
  return [
    2,
    requestAnimationFrame(callback)
  ];
};
export const Defer = (callback)=>{
  const scheduled = [
    2,
    0
  ];
  scheduled[1] = requestAnimationFrame(()=>{
    scheduled[0] = 0;
    scheduled[1] = setTimeout(callback, 0);
  });
  return scheduled;
};
export const IsScheduled = (value)=>{
  return Array.isArray(value) && (value.length === 2 || value.length === 3) && typeof value[0] === 'number' && typeof value[1] === 'number' && (value[2] === undefined || value[2] === true);
};
