// @spicemod/creator/dist/bin.mjs
#!/usr/bin/env node
import { Command, Option } from "commander";
import * as v from "valibot";
import path, { basename, dirname, extname, join, relative, resolve } from "node:path";
import { build, context, formatMessages, transform } from "esbuild";
import fs, { createReadStream, existsSync, lstatSync, mkdirSync, readFileSync, readdirSync, rmSync, statSync, writeFileSync } from "node:fs";
import { watchConfig } from "c12";
import { globSync } from "tinyglobby";
import { URL as URL$1, fileURLToPath } from "node:url";
import { execSync, spawn, spawnSync } from "node:child_process";
import readline, { createInterface } from "node:readline";
import * as p from "@clack/prompts";
import { cancel, log, spinner } from "@clack/prompts";
import pc from "picocolors";
import "dotenv/config";
import { readFile, unlink, writeFile } from "node:fs/promises";
import { parse } from "ini";
import { gzipSync } from "node:zlib";
import postcssMinify from "@csstools/postcss-minify";
import autoprefixer from "autoprefixer";
import { postcssModules, sassPlugin } from "esbuild-sass-plugin";
import postcss from "postcss";
import postcssImport from "postcss-import";
import postcssPresetEnv from "postcss-preset-env";
import { createHash } from "node:crypto";
import { chdir } from "node:process";
import { lookup } from "node:dns/promises";
import { mkdir, writeFile as writeFile$1 } from "fs/promises";
import { dirname as dirname$1, join as join$1 } from "path";
import { createServer } from "node:http";
import { createServer as createServer$1 } from "node:net";
import { WebSocket, WebSocketServer } from "ws";

//#region src/utils/replace.ts
function replace(contents, kv) {
	const keys = Object.keys(kv);
	if (keys.length === 0) return contents;
	const pattern = new RegExp(keys.map((k) => k.replace(/[.*+?^${}()|[\]\\]/g, "\\$&")).join("|"), "g");
	return contents.replace(pattern, (matched) => kv[matched] ?? matched);
}

//#endregion
//#region src/utils/fs.ts
function mkdirp(dir) {
	try {
		mkdirSync(dir, { recursive: true });
	} catch (e) {
		if (e?.code === "EEXIST") return;
		throw e;
	}
}
function dist(path, url = import.meta.url) {
	const insideDistFolder = url.includes("dist");
	return fileURLToPath(new URL(`./${!insideDistFolder ? "dist/" : ""}${path}`, url).href);
}

//#endregion
//#region src/metadata.ts
const toOptions = (metadata) => metadata.map((m) => ({
	value: m.name,
	label: m.title,
	hint: m.description
}));
const templateTypes = [
	"extension",
	"theme",
	"custom-app"
];
const templates = templateTypes.map((dir) => {
	const meta_file = dist(`templates/${dir}/meta.json`, import.meta.url);
	const { title, description } = JSON.parse(readFileSync(meta_file, "utf8"));
	return {
		name: dir,
		title,
		description
	};
});
const templateOptions = toOptions(templates);
const linterTypes = [
	"biome",
	"eslint",
	"oxlint",
	"none"
];
const linterMeta = {
	oxlint: {
		title: "Oxlint",
		description: "high-performance linter - http://oxc.rs"
	},
	biome: {
		title: "Biome",
		description: "Fast formatter and linter - https://biomejs.dev"
	},
	eslint: {
		title: "ESLint",
		description: "Industry standard, highly extensible - https://eslint.org"
	},
	none: {
		title: "None",
		description: "Skip linting for this project"
	}
};
const linters = linterTypes.map((name) => ({
	name,
	...linterMeta[name]
}));
const linterOptions = toOptions(linters);
const languageTypes = ["ts", "js"];
const languageMeta = {
	ts: {
		title: "Typescript",
		description: "with types (Recommended)"
	},
	js: {
		title: "Javascript",
		description: "with JSDOC"
	}
};
const languages = languageTypes.map((name) => ({
	name,
	...languageMeta[name]
}));
const languageOptions = toOptions(languages);
const frameworkTypes = ["react", "vanilla"];
const frameworkMeta = {
	react: {
		title: "ReactJS",
		description: "js framework - https://react.dev"
	},
	vanilla: {
		title: "Vanilla",
		description: "ya know it"
	}
};
const frameworks = frameworkTypes.map((name) => ({
	name,
	...frameworkMeta[name]
}));
const frameworkOptions = toOptions(frameworks);
const liveReloadFilePath = dist(`templates/liveReload.js`, import.meta.url);
const hmrCustomAppFilePath = dist(`templates/hmrCustomApp.js`, import.meta.url);
const templateWrapperFilePath = dist("templates/wrapper.js", import.meta.url);
const customAppEntryFilePath = dist("templates/customAppEntry.js", import.meta.url);
const injectCSSFilePath = dist("templates/injectCSS.js", import.meta.url);

//#endregion
//#region package.json
var name = "@spicemod/creator";
var version = "0.0.35";

//#endregion
//#region src/utils/common.ts
const runCommand = async (action, options = {
	outro: false,
	message: void 0
}) => {
	try {
		p.intro(options.message === void 0 || typeof options.message !== "function" ? `${name} ${pc.gray(`(v${version})`)}` : options.message({
			name,
			version
		}));
		await action();
		if (options.outro) p.outro("You're all set!");
	} catch (e) {
		if (e instanceof Error) {
			p.log.error(e.stack ?? String(e));
			p.log.message();
		}
		p.cancel("Operation failed.");
	}
};
const normalizePosix = (dir) => path.posix.normalize(dir.replace(/\\/g, "/"));
function urlSlugify(text) {
	return text.toString().normalize("NFD").replace(/[\u0300-\u036f]/g, "").toLowerCase().trim().replace(/\s+/g, "-").replace(/[^\w-]+/g, "").replace(/--+/g, "-");
}
function varSlugify(text) {
	return text.toString().normalize("NFD").replace(/[\u0300-\u036f]/g, "").toLowerCase().trim().replace(/\s+/g, "_").replace(/[^\w]/g, "").replace(/__+/g, "_").replace(/^[0-9]/, (val) => `_${val}`);
}

//#endregion
//#region src/utils/package-manager.ts
const packageManagers = [
	"npm",
	"pnpm",
	"yarn",
	"bun"
];
function getPackageManager() {
	const userAgent = process.env.npm_config_user_agent ?? "";
	if (userAgent.startsWith("yarn")) return "yarn";
	if (userAgent.startsWith("pnpm")) return "pnpm";
	if (userAgent.startsWith("bun")) return "bun";
	return "npm";
}
/**
* Spawn a package manager installation based on user preference.
*
* @returns A Promise that resolves once the installation is finished.
*/
async function installPackages(packageManager, isOnline) {
	const args = ["install"];
	if (!isOnline) {
		log.warn(pc.yellow("You appear to be offline.\nFalling back to the local cache."));
		args.push("--offline");
	}
	/**
	* Return a Promise that resolves once the installation is finished.
	*/
	return new Promise((resolve, reject) => {
		/**
		* Spawn the installation process.
		*/
		const child = spawn(packageManager, args, {
			stdio: [
				"inherit",
				"pipe",
				"pipe"
			],
			env: {
				...process.env,
				ADBLOCK: "1",
				NODE_ENV: "development",
				DISABLE_OPENCOLLECTIVE: "1"
			}
		});
		createInterface({ input: child.stdout }).on("line", (line) => {
			log.message(line);
		});
		createInterface({ input: child.stderr }).on("line", (line) => {
			log.error(line);
		});
		child.on("close", (code) => {
			log.message();
			if (code !== 0) {
				reject({ command: `${packageManager} ${args.join(" ")}` });
				return;
			}
			resolve();
		});
	});
}

//#endregion
//#region src/constants.ts
const GITHUB_LINK = "https://github.com/sanoojes/spicetify-creator";
const DISCORD_LINK = "https://discord.gg/YGkktjdYV8";
const DOCS_LINK = "https://github.com/sanoojes/spicetify-creator";
const CONFIG_REF_LINK = "https://github.com/sanoojes/spicetify-creator";
const SPICETIFY_LINK = "https://spicetify.app/docs/getting-started";
const DEV_MODE_VAR_NAME = `__SPICE_CREATOR_DEV__`;
const CHECK = pc.bold(pc.green("✔"));
const CROSS = pc.bold(pc.red("✖"));
const WARN = pc.bold(pc.yellow("⚠"));
const VALID_PROJECT_FILES = new Set([
	".DS_Store",
	".git",
	".gitattributes",
	".gitignore",
	".gitlab-ci.yml",
	".hg",
	".hgcheck",
	".hgignore",
	".idea",
	".npmignore",
	".travis.yml",
	"LICENSE",
	"Thumbs.db",
	"docs",
	"mkdocs.yml",
	"npm-debug.log",
	"yarn-debug.log",
	"yarn-error.log",
	"yarnrc.yml",
	".yarn"
]);
const CUSTOM_APP_NAME_LOCALES = [
	"ms",
	"gu",
	"ko",
	"pa-IN",
	"az",
	"ru",
	"uk",
	"nb",
	"sv",
	"sw",
	"ur",
	"bho",
	"pa-PK",
	"te",
	"ro",
	"vi",
	"am",
	"bn",
	"en",
	"id",
	"bg",
	"da",
	"es-419",
	"mr",
	"ml",
	"th",
	"tr",
	"is",
	"fa",
	"or",
	"he",
	"hi",
	"zh-TW",
	"sr",
	"pt-BR",
	"zu",
	"nl",
	"es",
	"lt",
	"ja",
	"st",
	"it",
	"el",
	"pt-PT",
	"kn",
	"de",
	"fr",
	"ne",
	"ar",
	"af",
	"et",
	"pl",
	"ta",
	"sl",
	"pk",
	"hr",
	"sk",
	"fi",
	"lv",
	"fil",
	"fr-CA",
	"cs",
	"zh-CN",
	"hu"
];

//#endregion
//#region src/config/schema.ts
const ServerConfigSchema = v.object({
	port: v.optional(v.number()),
	serveDir: v.optional(v.string()),
	hmrPath: v.optional(v.string())
});
const EntryFileSchema = v.string();
const AssetEntrySchema = v.object({
	js: EntryFileSchema,
	css: EntryFileSchema
});
const CustomAppEntrySchema = v.object({
	extension: EntryFileSchema,
	app: EntryFileSchema
});
const LocaleNameSchema = v.intersect([v.object({ en: v.string() }), v.record(v.picklist(CUSTOM_APP_NAME_LOCALES), v.string())]);
const ExtensionTemplateSchema = v.object({
	name: v.string(),
	template: v.literal("extension"),
	cssId: v.optional(v.string()),
	entry: EntryFileSchema
});
const ThemeTemplateSchema = v.object({
	name: v.string(),
	template: v.literal("theme"),
	entry: AssetEntrySchema
});
const CustomAppTemplateSchema = v.object({
	name: v.union([v.string(), LocaleNameSchema], "Name must be a string or a translations object containing the required 'en' locale."),
	icon: v.object({
		default: v.string(),
		active: v.optional(v.string())
	}),
	template: v.literal("custom-app"),
	entry: CustomAppEntrySchema
});
const TemplateSpecificSchema = v.variant("template", [
	ExtensionTemplateSchema,
	ThemeTemplateSchema,
	CustomAppTemplateSchema
]);
const TemplateSpecificOptionalSchema = v.variant("template", [
	v.partial(ExtensionTemplateSchema),
	v.partial(ThemeTemplateSchema),
	v.partial(CustomAppTemplateSchema)
]);
const RequiredCommonSchema = v.object({
	outDir: v.string(),
	linter: v.picklist(linterTypes),
	framework: v.picklist(frameworkTypes),
	packageManager: v.picklist(packageManagers),
	esbuildOptions: v.record(v.string(), v.any()),
	serverConfig: v.partial(ServerConfigSchema),
	version: v.string()
});
const OptionalCommonSchema = v.object({ devModeVarName: v.optional(v.string()) });
const FileOptionsSchema = v.intersect([
	v.partial(RequiredCommonSchema),
	OptionalCommonSchema,
	TemplateSpecificOptionalSchema
]);
const OptionsSchema$1 = v.intersect([
	v.required(RequiredCommonSchema),
	OptionalCommonSchema,
	TemplateSpecificSchema
]);

//#endregion
//#region src/env.ts
const isDev = process.env.IS_DEV === "true";
const spicetifyBin = process.env.SPICETIFY_BIN || process.env.SPICE_BIN || "spicetify";
const skipSpicetify = process.env.SPICETIFY_SKIP === "true" || !!process.env.CI;
const env = {
	isDev,
	spicetifyBin,
	skipSpicetify
};

//#endregion
//#region src/utils/logger.ts
var Logger = class {
	constructor(prefix = "", mode = env.isDev ? "dev" : "prod") {
		this.prefix = prefix;
		this.mode = mode;
	}
	get isDev() {
		return this.mode === "dev";
	}
	getPrefix() {
		if (!this.prefix || !env.isDev) return null;
		return pc.dim(`[${this.prefix.toLowerCase()}]`);
	}
	add(fn, args) {
		const p = this.getPrefix();
		if (p) fn(p, ...args);
		else fn(...args);
	}
	greeting(msg = "") {
		const label = pc.bgBlue(pc.black(` ${name.toUpperCase()} `));
		console.log(`${label} ${pc.dim(`v${version}`)} ${msg}`);
	}
	info(...args) {
		this.add(console.info, args);
	}
	success(m) {
		console.log(`${CHECK} ${pc.green(m)}`);
	}
	warn(m) {
		console.log(`${WARN} ${pc.yellow(m)}`);
	}
	cwarn(...args) {
		this.add(console.warn, args);
	}
	error(...errors) {
		const formatted = [];
		for (const err of errors) if (err instanceof Error) {
			formatted.push(pc.red(err.message));
			if (this.isDev && err.stack) {
				const stack = err.stack.split("\n").slice(1).join("\n");
				formatted.push(pc.dim(stack));
			}
		} else formatted.push(err);
		this.add(console.error, formatted);
	}
	log(...args) {
		if (!this.isDev) return;
		this.add(console.log, args);
	}
	debug(...args) {
		if (!this.isDev) return;
		this.add((...a) => console.log(pc.magenta("[debug]"), ...a), args);
	}
	clear() {
		if (!process.stdout.isTTY || process.env.CI) return;
		if (this.isDev) {
			this.log("clear skipped");
			return;
		}
		readline.cursorTo(process.stdout, 0, 0);
		readline.clearScreenDown(process.stdout);
	}
};
const createLogger = (prefix = "", mode) => new Logger(prefix, mode);
const logger$2 = createLogger("common");

//#endregion
//#region src/utils/spicetify/schema.ts
const SpicetifyConfigSchema = v.object({
	Setting: v.object({
		spotify_path: v.string(),
		prefs_path: v.string(),
		inject_theme_js: v.string(),
		inject_css: v.string(),
		current_theme: v.string(),
		color_scheme: v.string(),
		always_enable_devtools: v.string()
	}),
	AdditionalOptions: v.object({
		experimental_features: v.string(),
		extensions: v.pipe(v.string(), v.transform((input) => input.split("|").filter(Boolean))),
		custom_apps: v.string()
	}),
	Backup: v.object({
		version: v.string(),
		with: v.string()
	})
});

//#endregion
//#region src/utils/spicetify/index.ts
function runSpice(args) {
	if (env.skipSpicetify) throw new Error("Spicetify operations are disabled in CI");
	validateSpicetify(env.spicetifyBin);
	return spawnSync(env.spicetifyBin, args, { encoding: "utf-8" });
}
const getCustomAppsDir = () => join(getSpiceDataPath(), "CustomApps");
const getExtensionDir = () => join(getSpiceDataPath(), "Extensions");
const getThemesDir = () => join(getSpiceDataPath(), "Themes");
async function getSpicetifyConfig() {
	if (env.skipSpicetify) throw new Error("Spicetify operations are disabled in CI");
	const { stdout, stderr, error } = runSpice(["path", "-c"]);
	if (error || stderr) throw new Error(`Failed to locate Spicetify config: ${stderr || error?.message}`);
	const rawConfig = parse(await readFile(stdout.trim(), "utf-8"));
	const result = v.safeParse(SpicetifyConfigSchema, rawConfig);
	if (result.success) return result.output;
	else throw new Error("Spicetify Config Validation Failed:", v.flatten(result.issues).nested);
}
function getSpiceDataPath() {
	if (env.skipSpicetify) throw new Error("Spicetify operations are disabled in CI");
	const { stdout, stderr, error } = runSpice(["path", "userdata"]);
	if (error || stderr) throw new Error(`Failed to locate Spicetify config: ${stderr || error?.message}`);
	return stdout.trim();
}
function validateSpicetify(bin) {
	if (env.skipSpicetify) return;
	const result = spawnSync(bin, ["--version"], { encoding: "utf-8" });
	if (result.error) throw result.error;
	if (result.status !== 0) throw new Error(`Invalid spicetify binary "${bin}": ${result.stderr || "unknown error"}`);
}

//#endregion
//#region src/config/globs.ts
const JS_EXTENSIONS = "{ts,tsx,js,jsx,mts,mjs,cts,cjs}";
const CSS_EXTENSIONS = "{css,scss,sass,less,styl,stylus,pcss,postcss}";
const withSrc = (dirs) => dirs.flatMap((dir) => [dir, `src/${dir}`]);
const createGlobs = (names, dirs, ext) => names.flatMap((name) => dirs.map((dir) => `${dir}${name}.${ext}`));
const JS_ENTRY_GLOBS = createGlobs(["app", "index"], withSrc([""]), JS_EXTENSIONS);
const JS_EXTENSION_ENTRY_GLOBS = createGlobs(["app", "index"], withSrc(["extension/"]), JS_EXTENSIONS);
const CSS_ENTRY_GLOBS = createGlobs(["app"], withSrc([
	"",
	"styles/",
	"css/"
]), CSS_EXTENSIONS);
const ICON_GLOBS = createGlobs(["icon", "logo"], withSrc([
	"",
	"icons/",
	"assets/"
]), "svg");
const ICON_ACTIVE_GLOBS = createGlobs(["icon-active", "logo-active"], withSrc([
	"",
	"icons/",
	"assets/"
]), "svg");
const ENTRY_MAP = {
	js: [...JS_ENTRY_GLOBS, ...JS_EXTENSION_ENTRY_GLOBS],
	css: CSS_ENTRY_GLOBS,
	"js-app-only": JS_ENTRY_GLOBS,
	"js-extension-only": JS_EXTENSION_ENTRY_GLOBS
};

//#endregion
//#region src/config/index.ts
const logger$1 = createLogger("config");
let previousConfig;
const CONFIG_DEFAULTS = {
	outDir: "./dist",
	linter: "biome",
	framework: "react",
	template: "extension",
	packageManager: getPackageManager()
};
async function loadConfig(cb) {
	let cleanup;
	const runCb = async (config, isUpdate) => {
		if (isUpdate && previousConfig) await cleanupSpicetifyConfig();
		if (typeof cleanup === "function") await cleanup();
		cleanup = await cb(config, isUpdate);
		previousConfig = config;
	};
	const watcher = await watchConfig({
		name: "spice",
		defaults: CONFIG_DEFAULTS,
		configFileRequired: true,
		packageJson: true,
		async onUpdate({ oldConfig, newConfig }) {
			try {
				logger$1.debug("Config update triggered", {
					oldConfig,
					newConfig
				});
				const resolved = await getResolvedConfig(newConfig.config, { exitOnError: false });
				logger$1.debug("Resolved config", { resolved });
				try {
					await runCb(resolved, true);
				} catch (error) {
					logger$1.error(error);
				}
			} catch (err) {
				logger$1.error(pc.red("Config validation failed, keeping previous configuration"));
				if (err instanceof Error) logger$1.error(pc.dim(" └─ ") + err.message);
			}
		}
	});
	await runCb(await getResolvedConfig(watcher.config), false);
	return watcher;
}
async function getResolvedConfig(config, { exitOnError = true } = {}) {
	const resolvedContext = await resolveContext(config);
	const result = v.safeParse(OptionsSchema$1, resolvedContext);
	if (result.success) return result.output;
	logger$1.error(pc.red("Failed to load configuration:"));
	result.issues.forEach((issue) => {
		const path = issue.path?.map((p) => p.key).join(".") || "input";
		logger$1.error(`${pc.dim(" └─")} ${pc.yellow(path)}: ${pc.white(issue.message)}`);
	});
	if (exitOnError) process.exit(1);
	throw new Error("Invalid configuration");
}
async function resolveContext(config) {
	const cwd = process.cwd();
	const pkg = config.name && config.version ? {} : getPackageMeta(cwd);
	config.name ||= pkg.name || basename(cwd);
	config.version ||= pkg.version || "0.0.1";
	config.entry ||= resolveConfigEntries(config.template, cwd);
	config.outDir = resolve(cwd, config.outDir || "./dist");
	config.esbuildOptions ??= {};
	config.serverConfig ??= {};
	if (config.template === "extension") config.cssId ??= `${varSlugify(getEnName(config.name))}_styles`;
	if (config.template === "custom-app") config.icon ??= {
		default: resolveDefaultIcon(cwd),
		active: resolveActiveIcon(cwd)
	};
	return config;
}
function getPackageMeta(cwd) {
	try {
		return JSON.parse(readFileSync(resolve(cwd, "package.json"), "utf-8"));
	} catch {
		return {};
	}
}
function resolveConfigEntries(template, cwd) {
	if (template === "theme") return {
		js: resolveDefaultEntry(cwd, "js"),
		css: resolveDefaultEntry(cwd, "css")
	};
	if (template === "custom-app") return {
		app: resolveDefaultEntry(cwd, "js-app-only"),
		extension: resolveDefaultEntry(cwd, "js-extension-only")
	};
	return resolveDefaultEntry(cwd, "js");
}
function resolveDefaultEntry(cwd, type) {
	const globs = ENTRY_MAP[type];
	for (const glob of globs) {
		const matches = globSync(glob, {
			cwd,
			absolute: true
		});
		if (matches.length > 0 && matches[0]) return matches[0];
	}
	const displayType = {
		js: "JavaScript/TypeScript",
		css: "CSS/Stylesheet",
		"js-app-only": "JavaScript/TypeScript (Custom App)",
		"js-extension-only": "JavaScript/TypeScript (Extension)"
	}[type];
	const expectedFiles = globs.map((g) => `  - ${g}`).join("\n");
	throw new Error(`No ${displayType} entry file found in your project.\nPlease create one of the following files:\n${expectedFiles}`);
}
function resolveDefaultIcon(cwd) {
	for (const glob of ICON_GLOBS) {
		const matches = globSync(glob, {
			cwd,
			absolute: true
		});
		if (matches.length > 0 && matches[0]) return matches[0];
	}
	const expectedFiles = ICON_GLOBS.map((g) => `  - ${g}`).join("\n");
	throw new Error(`No icon file found in your project.\nPlease create one of the following files:\n${expectedFiles}`);
}
function resolveActiveIcon(cwd) {
	for (const glob of ICON_ACTIVE_GLOBS) {
		const matches = globSync(glob, {
			cwd,
			absolute: true
		});
		if (matches.length > 0 && matches[0]) return matches[0];
	}
	return "";
}
const getEnName = (configName) => typeof configName === "string" ? configName : configName.en;
function getSpiceIdentifier(config) {
	if (config.template === "extension") return `${urlSlugify(config.name)}.js`;
	if (config.template === "custom-app") return urlSlugify(getEnName(config.name));
	return urlSlugify(getEnName(config.name));
}
function getSpiceVarName(template) {
	if (template === "extension") return "extensions";
	if (template === "custom-app") return "custom_apps";
	return "current_theme";
}
async function cleanupSpicetifyConfig() {
	if (!previousConfig) return;
	const prevIdentifier = getSpiceIdentifier(previousConfig);
	const varName = getSpiceVarName(previousConfig.template);
	logger$1.debug(`Cleaning up previous spicetify config: ${varName} ${prevIdentifier}-`);
	runSpice([
		"config",
		varName,
		`${prevIdentifier}-`
	]);
	runSpice(["apply"]);
}

//#endregion
//#region src/esbuild/format.ts
function formatBuildSummary(files) {
	try {
		const tableData = [];
		let totalSize = 0;
		for (const [filePath, file] of files) {
			let gzipKB = 0;
			if (file.contents) gzipKB = gzipSync(file.contents).length / 1024;
			const sizeKB = file.contents.length / 1024;
			totalSize += file.contents.length;
			tableData.push({
				fullPath: relative(process.cwd(), filePath),
				sizeKB,
				gzipKB
			});
		}
		const maxPathWidth = Math.max(...tableData.map((d) => d.fullPath.length), 10);
		const lines = [];
		for (const file of tableData) {
			const sizeColor = file.sizeKB > 500 ? pc.magenta : file.sizeKB > 100 ? pc.yellow : pc.cyan;
			const pathPart = pc.dim(file.fullPath);
			const padding = " ".repeat(maxPathWidth - file.fullPath.length + 2);
			const sizeStr = `${file.sizeKB.toFixed(2)} kB`.padStart(9);
			const gzipStr = pc.dim(`│ gzip: ${file.gzipKB.toFixed(2).padStart(5)} kB`);
			lines.push(`${pc.blue("i")} ${pathPart}${padding}${sizeColor(sizeStr)} ${gzipStr}`);
		}
		const totalSizeKB = totalSize / 1024;
		const fileCount = tableData.length;
		lines.push(`${pc.blue("ℹ")} ${pc.bold(pc.gray(`${fileCount} files, total: ${totalSizeKB.toFixed(2)} kB`))}`);
		return lines.join("\n");
	} catch (e) {
		logger$2.debug(e);
		return pc.red("Failed to generate build summary");
	}
}

//#endregion
//#region src/esbuild/plugins/buildErrorReporter.ts
function formatErrorForFrontend(errors) {
	return errors.map((err) => {
		return {
			id: err.id,
			text: err.text,
			location: err.location ? {
				...err.location,
				lineText: err.location.lineText
			} : null,
			notes: err.notes,
			pluginName: err.pluginName,
			detail: err.detail
		};
	});
}
async function formatErrorForConsole(errors) {
	return (await formatMessages(errors, {
		kind: "error",
		color: true
	})).join("\n");
}
function buildErrorReporter({ server }) {
	return {
		name: "spice_internal__build-error-reporter",
		setup(build) {
			build.onEnd(async (result) => {
				const errors = result.errors;
				const warnings = result.warnings;
				if (errors.length === 0) {
					server?.broadcast({
						type: "build-success",
						errors: [],
						warnings
					});
					return;
				}
				const formattedErrors = formatErrorForFrontend(errors);
				logger$2.debug(pc.cyan("[buildErrorReporter] Broadcasting error to WebSocket clients"));
				server?.broadcast({
					type: "build-error",
					errors: formattedErrors,
					warnings
				});
			});
		}
	};
}
async function getFormattedErrors(errors) {
	return formatErrorForConsole(errors);
}

//#endregion
//#region src/esbuild/plugins/buildLogger.ts
const buildLogger = ({ cache }) => ({
	name: "spice_internal__build-logger",
	setup(build) {
		let isFirstBuild = true;
		let buildStartTime;
		build.onStart(() => {
			buildStartTime = performance.now();
			if (!isFirstBuild) {
				logger$2.clear();
				logger$2.info(pc.dim("Rebuilding..."));
			} else logger$2.info(pc.dim("Build started..."));
		});
		build.onEnd(async (result) => {
			if (result.errors.length > 0) {
				const formattedErrors = await getFormattedErrors(result.errors);
				logger$2.error(pc.red("Build failed.\n") + formattedErrors);
				return;
			}
			let moduleCount = 0;
			try {
				if (result.metafile) {
					result.metafile.outputs;
					moduleCount = Object.keys(result.metafile.inputs).length;
					const details = formatBuildSummary(cache.files);
					logger$2.info(details);
				}
			} catch {}
			logger$2.info(`${CHECK} ${moduleCount} modules transformed.`);
			logger$2.info(pc.green(`${CHECK} built in ${getTime(buildStartTime)}.`));
			isFirstBuild = false;
		});
	}
});
function getTime(start) {
	const ms = performance.now() - start;
	return ms > 1e3 ? `${(ms / 1e3).toFixed(2)}s` : `${Math.round(ms)}ms`;
}

//#endregion
//#region src/esbuild/plugins/clean.ts
const clean = (cache, logger = createLogger("plugin:clean")) => ({
	name: "clean",
	setup(build) {
		const outdir = resolve(build.initialOptions.outdir || "./dist");
		build.onStart(() => {
			if (fs.existsSync(outdir)) try {
				fs.rmSync(outdir, {
					recursive: true,
					force: true
				});
				logger.log(`Removed: ${outdir}`);
			} catch (err) {
				logger.error(`Error removing ${outdir}:`, err);
			}
		});
	}
});

//#endregion
//#region src/esbuild/plugins/inlineBundledCss.ts
function inlineBundledCss(styleId) {
	return {
		name: "inline-bundled-css",
		setup(build) {
			build.onEnd(async (result) => {
				if (result.errors.length > 0 || !result.outputFiles) return;
				const outputFiles = result.outputFiles;
				const cssFiles = outputFiles.filter((file) => file.path.endsWith(".css"));
				const jsFiles = outputFiles.filter((file) => file.path.endsWith(".js"));
				await Promise.all(cssFiles.map(async (cssFile) => {
					const cssContent = cssFile.text;
					const { code: injectScript } = await transform(readFileSync(injectCSSFilePath, "utf-8"), {
						loader: "js",
						define: {
							__ESBUILD__STYLE_ID__: JSON.stringify(styleId),
							__ESBUILD__CSS_CONTENT__: JSON.stringify(cssContent)
						}
					});
					const jsPath = cssFile.path.replace(/\.css$/, ".js");
					const jsFile = jsFiles.find((file) => file.path === jsPath);
					if (jsFile) {
						const updatedJsText = jsFile.text + injectScript;
						jsFile.contents = new TextEncoder().encode(updatedJsText);
						const cssIndex = outputFiles.indexOf(cssFile);
						if (cssIndex > -1) outputFiles.splice(cssIndex, 1);
					}
				}));
			});
		}
	};
}

//#endregion
//#region src/esbuild/plugins/css.ts
function css({ minify = false, inline = false, logger = createLogger("plugin:css"), styleId = null } = {}) {
	const postCssPlugins = [
		postcssImport({ path: [resolve(process.cwd(), "src")] }),
		autoprefixer,
		postcssPresetEnv({ stage: 0 }),
		...minify ? [postcssMinify()] : []
	];
	const type = inline ? styleId ? "css" : "style" : "css";
	const plugins = [sassPlugin({
		filter: /\.module\.(s[ac]ss|css)$/,
		type,
		transform: postcssModules({
			getJSON: () => {},
			generateScopedName: "[name]__[local]___[hash:base64:5]"
		}, postCssPlugins)
	}), sassPlugin({
		filter: /\.(s[ac]ss|css)$/,
		type,
		async transform(css, _resolveDir, filePath) {
			const start = performance.now();
			logger.log("processing:", filePath, "type:", type);
			const result = await postcss(postCssPlugins).process(css, { from: filePath });
			logger.debug("Global CSS processed", {
				filePath,
				ms: Math.round(performance.now() - start)
			});
			return result.css;
		}
	})];
	if (inline && styleId) plugins.push(inlineBundledCss(styleId));
	return plugins;
}

//#endregion
//#region src/esbuild/plugins/externalGlobal.ts
const externalGlobal = (externals, namespace = "spicetify-global") => {
	return {
		name: "spice_internal__external-global",
		setup(build) {
			build.onResolve({ filter: new RegExp(`^(${Object.keys(externals).join("|")})$`) }, (args) => ({
				path: args.path,
				namespace
			}));
			build.onLoad({
				filter: /.*/,
				namespace
			}, (args) => {
				return { contents: `module.exports = ${externals[args.path]}` };
			});
		}
	};
};

//#endregion
//#region src/esbuild/plugins/spicetifyHandlers.ts
async function removeFile(logger, targetPath) {
	try {
		await unlink(targetPath);
		logger.debug(pc.green(`${CHECK} Removed: ${targetPath}`));
	} catch (err) {
		if (err.code !== "ENOENT") logger.error(pc.red(`${CROSS} Failed to remove file: ${err instanceof Error ? err.message : String(err)}`));
	}
}
async function copyFiles(logger, destDirs, cacheFiles) {
	const tasks = [];
	for (const filePath of cacheFiles.keys()) {
		const fileData = cacheFiles.get(filePath);
		if (!fileData) continue;
		for (const destDir of destDirs) {
			const targetPath = resolve(destDir, basename(filePath));
			tasks.push((async () => {
				await mkdirp(destDir);
				await writeFile(targetPath, fileData.contents);
			})());
		}
	}
	await Promise.all(tasks);
}
async function removeDeletedFiles(logger, destDirs, removedFiles) {
	const tasks = [];
	for (const removedPath of removedFiles) for (const destDir of destDirs) {
		const targetPath = resolve(destDir, basename(removedPath));
		tasks.push(removeFile(logger, targetPath));
	}
	await Promise.all(tasks);
}
const spicetifyHandler = ({ config, options, cache, logger = createLogger("plugin:spicetify-handler") }) => ({
	name: "spice_internal__spicetify-build-handler",
	async setup(build) {
		const { apply = true, copy = true, applyOnce = true, remove, outDir = "./dist" } = options;
		let hasAppliedOnce = false;
		const isExtension = config.template === "extension";
		const isCustomApp = config.template === "custom-app";
		const identifier = isExtension ? `${urlSlugify(config.name)}.js` : urlSlugify(getEnName(config.name));
		const getDestDirs = () => {
			const dirs = [resolve(outDir)];
			if (copy && !env.skipSpicetify) dirs.push(isExtension ? getExtensionDir() : isCustomApp ? resolve(getCustomAppsDir(), identifier) : resolve(getThemesDir(), identifier));
			return dirs;
		};
		if (env.skipSpicetify) logger.info(pc.yellow("Skipping Spicetify operations..."));
		else {
			const spiceConfig = await getSpicetifyConfig();
			logger.debug(pc.green("Spicetify Config: "), spiceConfig);
			if (apply) {
				const defaultTheme = spiceConfig?.Setting?.current_theme || "SpicetifyDefault";
				const spiceIdentifier = remove ? `${identifier}-` : identifier;
				runSpice([
					"config",
					isExtension ? "extensions" : isCustomApp ? "custom_apps" : "current_theme",
					spiceIdentifier
				]);
				if (!isExtension && !remove) {
					const cleanup = () => {
						if (isCustomApp) runSpice([
							"config",
							"custom_apps",
							`${identifier}-`
						]);
						else runSpice([
							"config",
							"current_theme",
							defaultTheme
						]);
						process.exit();
					};
					process.once("SIGINT", cleanup);
					process.once("SIGTERM", cleanup);
				}
			}
		}
		build.onEnd(async (result) => {
			if (result.errors.length > 0) return;
			const destDirs = getDestDirs();
			try {
				await copyFiles(logger, destDirs, cache.files);
				logger.debug(pc.green(`${CHECK} Built files written to ${outDir}`));
			} catch (err) {
				logger.error(pc.red(`${CROSS} Failed to copy files: ${err instanceof Error ? err.message : String(err)}`));
				return;
			}
			if (cache.removed.size > 0) {
				await removeDeletedFiles(logger, destDirs, cache.removed);
				cache.hasChanges = true;
			}
			cache.removed.clear();
			if (env.skipSpicetify) return;
			if (apply && cache.hasChanges && (!applyOnce || !hasAppliedOnce)) {
				const { stdout, stderr, status } = runSpice(["apply"]);
				if (status !== 0) logger.error(pc.red(`${CROSS} Spicetify apply failed:`), stdout, stderr);
				else hasAppliedOnce = true;
			}
		});
	}
});

//#endregion
//#region src/esbuild/plugins/wrapWithLoader.ts
function wrapWithLoader({ config, cache, outFiles, server, dev = false, logger = createLogger("plugin:wrapper") }) {
	const namespace = "spice_internal__wrap-with-loader";
	const name = typeof config.name === "string" ? config.name : config.name.en;
	const { template: type, version } = config;
	let previousManifestHash;
	return {
		name: namespace,
		setup(build$3) {
			if (build$3.initialOptions.write !== false) throw new Error(`[${namespace}] This plugin requires "write: false" in build options.`);
			build$3.onStart(() => {
				cache.changed.clear();
				cache.removed.clear();
				cache.hasChanges = false;
			});
			build$3.onEnd(async (res) => {
				try {
					if (res.errors.length > 0 || !res.outputFiles) return;
					const filesChanged = [];
					const outdir = resolve(build$3.initialOptions.outdir || "./dist");
					const bundledCss = getBundledCss(res.outputFiles, outdir, type, dev);
					const minify = build$3.initialOptions.minify;
					const currentFilePaths = /* @__PURE__ */ new Set();
					for (const file of res.outputFiles) {
						const isJs = file.path.endsWith(".js");
						const isCss = file.path.endsWith(".css");
						if (!dev && isCss && type === "extension") continue;
						const relPath = file.path.slice(outdir.length);
						const isCustomAppExtension = type === "custom-app" && isExtensionDir(relPath);
						let targetName;
						if (isJs) targetName = isCustomAppExtension ? outFiles.jsExtension ?? "extension.js" : outFiles.js;
						else if (isCss && !isCustomAppExtension) targetName = outFiles.css;
						if (targetName) currentFilePaths.add(join(outdir, targetName));
					}
					for (const cachedPath of cache.files.keys()) if (!currentFilePaths.has(cachedPath)) cache.removed.add(cachedPath);
					const transformPromises = res.outputFiles.map(async (file) => {
						const isJs = file.path.endsWith(".js");
						const isCss = file.path.endsWith(".css");
						if (!dev && isCss && type === "extension") return;
						const relPath = file.path.slice(outdir.length);
						const isCustomAppExtension = type === "custom-app" && isExtensionDir(relPath);
						let targetName;
						if (isJs) targetName = isCustomAppExtension ? outFiles.jsExtension ?? "extension.js" : outFiles.js;
						else if (isCss && !isCustomAppExtension) targetName = outFiles.css;
						if (!targetName) {
							logger.debug("Skipped file: ", file.path);
							return;
						}
						const renamedPath = join(build$3.initialOptions.outdir || "./dist/", targetName);
						if (type === "custom-app" && targetName === outFiles.js && isJs) {
							const globalName = varSlugify(getEnName(config.name));
							const final = (await build({
								bundle: true,
								write: false,
								minify,
								platform: "browser",
								format: "iife",
								globalName,
								stdin: {
									contents: readFileSync(customAppEntryFilePath),
									loader: "tsx",
									sourcefile: "entry.tsx"
								},
								plugins: [plugins.externalGlobal({ react: "Spicetify.React" }), {
									name: "virtual-modules",
									setup(vBuild) {
										vBuild.onResolve({ filter: /^virtual:app$/ }, () => ({
											path: "app",
											namespace: "virtual"
										}));
										vBuild.onLoad({
											filter: /.*/,
											namespace: "virtual"
										}, () => ({
											contents: file.text,
											loader: "js"
										}));
									}
								}]
							})).outputFiles?.[0];
							if (!final) return;
							let combinedCode = `${final.text}${dev ? `export default () => ${globalName}.default();` : `var render = () => ${globalName}.default();`}\n`;
							const nextBuffer = Buffer.from(combinedCode);
							const existingFile = cache.files.get(renamedPath);
							const nextHash = final.hash;
							if (!existingFile || existingFile.hash !== nextHash || config.template === "custom-app") {
								cache.files.set(renamedPath, {
									contents: nextBuffer,
									name: targetName,
									hash: final.hash
								});
								cache.changed.add(renamedPath);
								cache.hasChanges = true;
								filesChanged.push(renamedPath);
							}
							return;
						}
						if (!isJs) {
							const nextBuffer = Buffer.from(file.contents);
							const existingFile = cache.files.get(renamedPath);
							const nextHash = file.hash;
							if (!existingFile || existingFile.hash !== nextHash || config.template === "custom-app") {
								cache.files.set(renamedPath, {
									name: targetName,
									contents: nextBuffer,
									hash: file.hash
								});
								cache.changed.add(renamedPath);
								cache.hasChanges = true;
								filesChanged.push(renamedPath);
							}
							return;
						}
						const { code: transformedTemp } = await transform(readFileSync(templateWrapperFilePath, "utf-8"), {
							minify,
							target: build$3.initialOptions.target || "es2020",
							loader: "jsx",
							define: {
								__ESBUILD__HAS_CSS: JSON.stringify(type !== "theme" && bundledCss.length !== 0),
								__ESBUILD__INJECTED_CSS: JSON.stringify(bundledCss),
								__ESBUILD__APP_TYPE: JSON.stringify(type),
								__ESBUILD__APP_ID: JSON.stringify(varSlugify(name)),
								__ESBUILD__APP_VERSION: JSON.stringify(version)
							}
						});
						const template = replace(transformedTemp, {
							"\"{{INJECT_START_COMMENT}}\"": minify ? "" : "/* --- START --- */",
							"\"{{INJECT_END_COMMENT}}\"": minify ? "" : "/* --- END --- */",
							"\"{{INJECTED_JS_HERE}}\"": file.text
						});
						const nextBuffer = Buffer.from(template);
						const existingFile = cache.files.get(renamedPath);
						const nextHash = file.hash;
						if (!existingFile || existingFile.hash !== nextHash || config.template === "custom-app") {
							cache.files.set(renamedPath, {
								name: targetName,
								contents: nextBuffer,
								hash: file.hash
							});
							cache.changed.add(renamedPath);
							cache.hasChanges = true;
							filesChanged.push(renamedPath);
						}
					});
					await Promise.all(transformPromises);
					if (type === "custom-app") {
						const icon = readFileSync(config.icon.default).toString();
						const activeIcon = config.icon.active ? readFileSync(config.icon.active).toString() : icon;
						const manifestPath = join(outdir, "manifest.json");
						const manifest = {
							name: config.name,
							subfiles: [],
							subfiles_extension: ["extension.js"],
							icon,
							activeIcon
						};
						const manifestString = JSON.stringify(manifest, null, 2);
						const currentHash = createHash("md5").update(manifestString).digest("hex");
						if (currentHash !== previousManifestHash) {
							previousManifestHash = currentHash;
							const manifestBuffer = Buffer.from(manifestString);
							cache.files.set(manifestPath, {
								contents: manifestBuffer,
								name: "manifest.json",
								hash: currentHash
							});
							cache.changed.add(manifestPath);
							cache.hasChanges = true;
							filesChanged.push(manifestPath);
						}
					}
					if (filesChanged.length > 0) server?.broadcast(filesChanged);
				} catch (e) {
					logger.error(`Error: ${e instanceof Error ? e.message : String(e)}`);
				}
			});
		}
	};
}
function isExtensionDir(relPath) {
	return relPath.startsWith("/extension/") || relPath.startsWith("\\extension\\");
}
function getBundledCss(files, outdir, type, dev) {
	const cssFiles = files.filter((f) => f.path.endsWith(".css"));
	if (!dev && type === "extension") return cssFiles.map((f) => f.text).join("");
	if (type === "custom-app") return cssFiles.filter((f) => isExtensionDir(f.path.slice(outdir.length))).map((f) => f.text).join("");
	return "";
}

//#endregion
//#region src/esbuild/plugins/index.ts
const plugins = {
	css,
	clean,
	buildLogger,
	buildErrorReporter,
	externalGlobal,
	wrapWithLoader,
	spicetifyHandler
};

//#endregion
//#region src/esbuild/index.ts
const defaultBuildOptions = {
	bundle: true,
	write: false,
	metafile: true,
	treeShaking: true,
	jsx: "transform",
	format: "esm",
	platform: "browser",
	target: ["es2022", "chrome120"]
};
const getCommonPlugins = (opts) => {
	const { template, minify, buildOptions, outFiles, server, dev } = opts;
	const inline = !dev && template === "extension";
	const cache = {
		files: /* @__PURE__ */ new Map(),
		changed: /* @__PURE__ */ new Set(),
		removed: /* @__PURE__ */ new Set(),
		hasChanges: true
	};
	const p = [
		...plugins.css({
			minify,
			inline,
			styleId: template === "extension" && opts.cssId ? opts.cssId : null
		}),
		plugins.clean(cache),
		plugins.externalGlobal({
			react: "Spicetify.React",
			"react-dom": "Spicetify.ReactDOM",
			"react-dom/client": "Spicetify.ReactDOM",
			"react-dom/server": "Spicetify.ReactDOMServer",
			"react/jsx-runtime": "Spicetify.ReactJSX"
		}),
		plugins.wrapWithLoader({
			config: opts,
			cache,
			outFiles,
			server,
			dev
		}),
		plugins.spicetifyHandler({
			config: opts,
			cache,
			options: buildOptions
		}),
		plugins.buildLogger({ cache })
	];
	if (dev) p.push(plugins.buildErrorReporter({ server }));
	return p;
};
function getEntryPoints(config) {
	if (config.template === "theme") return [config.entry.js, config.entry.css];
	if (config.template === "custom-app") return [config.entry.app, config.entry.extension];
	return [config.entry];
}
function getOutFiles(config, isDev = false) {
	switch (config.template) {
		case "custom-app": return {
			js: "index.js",
			css: "style.css",
			jsExtension: "extension.js",
			manifest: "manifest.json"
		};
		case "extension": return {
			js: `${urlSlugify(getEnName(config.name))}.js`,
			css: isDev ? "app.css" : void 0
		};
		case "theme": return {
			js: "theme.js",
			css: "user.css"
		};
		default: throw new Error("Unknown template");
	}
}

//#endregion
//#region src/build/index.ts
const logger = createLogger("build");
async function build$2(options) {
	logger.clear();
	logger.greeting(pc.green("Building for production..."));
	let ctx;
	await loadConfig(async (config, isNewUpdate) => {
		if (isNewUpdate) {
			logger.clear();
			logger.info(pc.green("Config updated, reloading..."));
		}
		if (ctx) await ctx.dispose();
		ctx = await context(getJSBuildOptions(config, options));
		if (options.watch) {
			logger.info(pc.blue("Watching for changes..."));
			await ctx.watch();
		} else try {
			await ctx.rebuild();
		} catch {} finally {
			await ctx.dispose();
			process.exit(0);
		}
		return async () => {
			await ctx?.dispose();
			ctx = void 0;
		};
	});
}
function getJSBuildOptions(config, options) {
	const entryPoints = getEntryPoints(config);
	const minify = options.watch ? false : options.minify;
	const outDir = resolve(config.outDir);
	const outFiles = getOutFiles(config);
	const overrides = {
		...defaultBuildOptions,
		outdir: outDir,
		format: "esm",
		minify,
		globalName: varSlugify(getEnName(config.name)),
		sourcemap: false,
		external: [
			...config.esbuildOptions?.external ? config.esbuildOptions.external : [],
			"react",
			"react-dom"
		],
		define: {
			[DEV_MODE_VAR_NAME]: "false",
			...config.devModeVarName ? { [config.devModeVarName]: "false" } : {},
			...config.esbuildOptions.define
		},
		plugins: [...config.esbuildOptions?.plugins ? config.esbuildOptions.plugins : [], ...getCommonPlugins({
			...config,
			minify,
			buildOptions: {
				apply: options.apply,
				copy: options.copy,
				outDir,
				applyOnce: false,
				remove: false
			},
			outFiles
		})]
	};
	return {
		entryPoints,
		...config.esbuildOptions,
		...overrides
	};
}

//#endregion
//#region src/utils/schema.ts
function safeParse(schema, data, type = "CLI") {
	const result = v.safeParse(schema, data);
	if (result.success) return result.output;
	logger$2.error(`\n${pc.bgRed(pc.black(" ERROR "))} ${pc.red(`Invalid ${type} options:`)}`);
	result.issues.forEach((issue) => {
		const path = issue.path?.map((p) => p.key).join(".") || "input";
		logger$2.error(`${pc.dim(" └─")} ${pc.yellow(path)}: ${pc.white(issue.message)}`);
	});
	logger$2.error(`\n${pc.dim("Check your command flags and try again.")}\n`);
	process.exit(1);
}

//#endregion
//#region src/commands/build.ts
const CLIOptionsSchema$1 = v.strictObject({
	watch: v.boolean(),
	minify: v.boolean(),
	apply: v.boolean(),
	copy: v.boolean()
});
const build$1 = new Command("build").description("Build your spicetify project").option("-a, --apply", "Apply to spicetify", false).option("-w, --watch", "Watch mode", false).option("--no-copy", "Do not copy files to spicetify").option("--no-minify", "Disable code minification").action(async (opts) => {
	await build$2(safeParse(CLIOptionsSchema$1, opts));
});

//#endregion
//#region src/create/package.ts
const FRAMEWORKS$1 = {
	react: {
		dependencies: {
			react: "^18.3.1",
			"react-dom": "^18.3.1"
		},
		devDependencies: {
			"@types/react": "^18.3.27",
			"@types/react-dom": "^18.3.7"
		}
	},
	vanilla: {}
};
const LINTERS$1 = {
	eslint: {
		scripts: { lint: "eslint ." },
		devDependencies: {
			eslint: "^9.39.2",
			"@eslint/js": "^9.39.2",
			"@eslint/css": "^0.14.1",
			globals: "^17.2.0",
			jiti: "^2.6.1"
		}
	},
	biome: {
		scripts: { lint: "biome check --apply ." },
		devDependencies: { "@biomejs/biome": "latest" }
	},
	oxlint: {
		scripts: {
			lint: "oxlint",
			"lint:fix": "oxlint --fix"
		},
		devDependencies: { oxlint: "^1.43.0" }
	}
};
const INTERSECTIONS = { reactEslint: {
	condition: (options) => options.framework === "react" && options.linter === "eslint",
	devDependencies: {
		"eslint-plugin-react": "^7.37.5",
		"typescript-eslint": "^8.54.0"
	}
} };
function createPackageJSON(options) {
	const result = {
		name: options.name,
		description: `A Spicetify ${options.template} created with @spicetify/creator`,
		version: "0.0.1",
		type: "module",
		private: true,
		scripts: {
			sc: "spicetify-creator",
			dev: "spicetify-creator dev",
			build: "spicetify-creator build",
			"update-types": "spicetify-creator update-types",
			"clean-spice": "spicetify-creator clean-spicetify --all"
		},
		dependencies: {},
		devDependencies: { "@spicetify/creator": "latest" }
	};
	if (options.language === "ts") {
		result.devDependencies = {
			...result.devDependencies,
			...options.packageManager === "bun" ? { "@types/bun": "^1.3.10" } : { "@types/node": "^25.5.0" }
		};
		result.peerDependencies = {
			...result.peerDependencies,
			typescript: "^5"
		};
	}
	const slices = [FRAMEWORKS$1[options.framework], LINTERS$1[options.linter]];
	Object.values(INTERSECTIONS).forEach((intersection) => {
		if (intersection.condition(options)) slices.push(intersection);
	});
	slices.forEach((slice) => {
		if (!slice) return;
		for (const key of [
			"scripts",
			"dependencies",
			"devDependencies"
		]) if (slice[key]) result[key] = {
			...result[key],
			...slice[key]
		};
	});
	return result;
}
function writePackageJSON(packageJSON, targetDir) {
	try {
		mkdirp(targetDir);
		const data = `${JSON.stringify(packageJSON, null, 2)}\n`;
		writeFileSync(join(targetDir, "package.json"), data, "utf8");
	} catch (e) {
		const message = e instanceof Error ? e.message : "Unknown error";
		log.error(`Failed to write package.json: ${message}`);
		cancel("Exiting...");
		process.exit(1);
	}
}
function validateProjectName(name) {
	if (!name.length) return "Project name is required";
	if (!/^(?:@[a-z0-9-*~][a-z0-9-*._~]*\/)?[a-z0-9-~][a-z0-9-._~]*$/.test(name)) return "Invalid project name (must be lowercase and URL-friendly)";
}

//#endregion
//#region src/create/template.ts
const ext = (lang, jsx = false) => lang === "ts" ? `ts${jsx ? "x" : ""}` : `js${jsx ? "x" : ""}`;
const kv = ({ name, language, framework, linter, packageManager, template }) => ({
	"{{project-name}}": name,
	"{{project-url}}": `/${urlSlugify(name)}`,
	"{{framework}}": framework,
	"{{linter}}": linter,
	"{{package-manager}}": packageManager,
	"{{template}}": template,
	"{{language}}": ext(language),
	"{{entry-ext}}": `${ext(language)}${framework === "react" ? "x" : ""}`,
	"{{docs-link}}": DOCS_LINK,
	"{{get-started-link}}": DOCS_LINK,
	"{{discord-link}}": DISCORD_LINK,
	"{{github-link}}": GITHUB_LINK,
	"{{spicetify-link}}": SPICETIFY_LINK,
	"{{config-reference-link}}": CONFIG_REF_LINK
});
const action = { modify(c, opts) {
	return replace(c, kv(opts));
} };
const SHARED_FILES = (opts) => {
	const isShared = true;
	const files = [
		{
			from: "README.template.md",
			to: "README.md",
			action,
			isShared
		},
		{
			from: "DOT-gitignore",
			to: ".gitignore",
			isShared
		},
		{
			from: `spice.config.${ext(opts.language)}`,
			to: `spice.config.${ext(opts.language)}`,
			action,
			isShared
		}
	];
	if (opts.template === "custom-app") {
		files.push({
			from: `css/app.module.scss`,
			to: `src/css/app.module.scss`,
			action,
			isShared
		});
		files.push({
			from: `icon.svg`,
			to: `src/icon.svg`,
			action,
			isShared
		});
		files.push({
			from: `icon-active.svg`,
			to: `src/icon-active.svg`,
			action,
			isShared
		});
		files.push({
			from: "app.css",
			to: "src/extension/app.css",
			action,
			isShared
		});
	} else files.push({
		from: "app.css",
		to: "src/app.css",
		action,
		isShared
	});
	return files;
};
const LANGUAGE_FILES = {
	js: [{
		from: "jsconfig.json",
		to: "jsconfig.json",
		isShared: true
	}],
	ts: [
		{
			from: "tsconfig.json",
			to: "tsconfig.json",
			isShared: true
		},
		{
			from: "tsconfig.app.json",
			to: "tsconfig.app.json",
			isShared: true
		},
		{
			from: "tsconfig.node.json",
			to: "tsconfig.node.json",
			isShared: true
		}
	]
};
const FRAMEWORKS = {
	react: ({ language, template }) => {
		const react = [{
			from: `src/app.${ext(language, true)}`,
			to: `src/app.${ext(language, true)}`,
			action
		}, {
			from: `src/components/Onboarding.${ext(language, true)}`,
			to: `src/components/Onboarding.${ext(language, true)}`,
			action
		}];
		if (template === "custom-app") react.push({
			from: `src/extension/index.${ext(language, true)}`,
			to: `src/extension/index.${ext(language, true)}`,
			action
		});
		return react;
	},
	vanilla: ({ language, template }) => {
		const vanilla = [{
			from: `src/app.${ext(language)}`,
			to: `src/app.${ext(language)}`,
			action
		}, {
			from: `src/components/Onboarding.${ext(language)}`,
			to: `src/components/Onboarding.${ext(language)}`,
			action
		}];
		if (template === "custom-app") throw new Error("vanilla doesn't exist for custom-app");
		return vanilla;
	}
};
const LINTERS = {
	biome: [{
		from: "biome.json",
		to: "biome.json",
		isShared: true
	}],
	eslint: ({ language }) => [{
		from: `eslint.config.${ext(language)}`,
		to: `eslint.config.${ext(language)}`
	}],
	oxlint: [{
		from: "DOT-oxlintrc.json",
		to: ".oxlintrc.json",
		isShared: true
	}]
};
const getFiles = (options) => {
	const resolve = (slice) => typeof slice === "function" ? slice(options) : slice ?? [];
	return [
		...resolve(SHARED_FILES),
		...resolve(LANGUAGE_FILES[options.language]),
		...resolve(FRAMEWORKS[options.framework]),
		...resolve(LINTERS[options.linter])
	];
};
function setupTemplateFiles(options, targetDir) {
	const { template, language, framework } = options;
	const templateRoot = dist(`templates/${template}`, import.meta.url);
	const fromDir = join(templateRoot, language, framework);
	for (const file of getFiles(options)) {
		const src = (() => {
			if (file.isGlobal) return join(templateRoot, file.from);
			if (file.isShared) return join(templateRoot, "shared", file.from);
			return join(fromDir, file.from);
		})();
		const dest = join(targetDir, file.to);
		if (!existsSync(src)) {
			log.warn(`[Template] Source missing: ${src}`);
			continue;
		}
		mkdirp(dirname(dest));
		let content = readFileSync(src, "utf8");
		if (file.action?.modify) content = file.action.modify(content, options);
		writeFileSync(dest, content, "utf8");
	}
}

//#endregion
//#region src/utils/is-online.ts
function getProxy() {
	if (process.env.https_proxy) return process.env.https_proxy;
	try {
		const httpsProxy = execSync("npm config get https-proxy").toString().trim();
		return httpsProxy !== "null" ? httpsProxy : void 0;
	} catch {
		return;
	}
}
async function getOnline() {
	try {
		await lookup("registry.yarnpkg.com");
		return true;
	} catch {
		const proxy = getProxy();
		if (!proxy) return false;
		let url;
		try {
			url = new URL$1(proxy);
		} catch {
			return false;
		}
		const { hostname } = url;
		if (!hostname) return false;
		try {
			await lookup(hostname);
			return true;
		} catch {
			return false;
		}
	}
}

//#endregion
//#region src/utils/git.ts
function isInGitRepository() {
	try {
		execSync("git rev-parse --is-inside-work-tree", { stdio: "ignore" });
		return true;
	} catch {}
	return false;
}
function isInMercurialRepository() {
	try {
		execSync("hg --cwd . root", { stdio: "ignore" });
		return true;
	} catch {}
	return false;
}
function isDefaultBranchSet() {
	try {
		execSync("git config init.defaultBranch", { stdio: "ignore" });
		return true;
	} catch {}
	return false;
}
function tryGitInit(root) {
	let didInit = false;
	try {
		execSync("git --version", { stdio: "ignore" });
		if (isInGitRepository() || isInMercurialRepository()) return false;
		execSync("git init", { stdio: "ignore" });
		didInit = true;
		if (!isDefaultBranchSet()) execSync("git checkout -b main", { stdio: "ignore" });
		execSync("git add -A", { stdio: "ignore" });
		execSync("git commit -m \"Initial commit from @spicetify/creator create\"", { stdio: "ignore" });
		return true;
	} catch {
		if (didInit) try {
			rmSync(join(root, ".git"), {
				recursive: true,
				force: true
			});
		} catch {}
		return false;
	}
}

//#endregion
//#region src/utils/update-types.ts
const downloads = { spicetify: {
	from: "https://raw.githubusercontent.com/spicetify/cli/main/globals.d.ts",
	to: "./src/types/spicetify.d.ts",
	action: (content) => content.replace("const React: any;", "const React: typeof import(\"react\");").replace("const ReactDOM: any;", "const ReactDOM: typeof import(\"react-dom/client\");").replace("const ReactJSX: any;", "const ReactJSX: typeof import(\"react/jsx-runtime\");").replace("const ReactDOMServer: any;", "const ReactDOMServer: typeof import(\"react-dom/server\");")
} };
async function updateTypes(isUpdating = true, cwd = process.cwd()) {
	const s = spinner();
	s.start(`${isUpdating ? "Updating" : "Creating"} Types...`);
	await Promise.all(Object.entries(downloads).map(([name, download]) => downloadFile(name, download, isUpdating, cwd)));
	s.stop(`${isUpdating ? "Updated" : "Created"} Types!`);
}
async function downloadFile(name, { from, to, action }, isUpdating, cwd) {
	try {
		const res = await fetch(from);
		if (!res.ok) throw new Error(`HTTP Error: ${res.status} ${res.statusText}`);
		let text = await res.text();
		if (action) text = await action(text);
		const fullPath = join$1(cwd, to);
		await mkdir(dirname$1(fullPath), { recursive: true });
		await writeFile$1(fullPath, text, "utf8");
		const actionLog = isUpdating ? "updated" : "created";
		logger$2.log(`${name}.d.ts ${actionLog} (${from} -> ${fullPath})`);
	} catch (e) {
		log.error(`${name} failed`);
		console.error(e);
	}
}

//#endregion
//#region src/create/index.ts
const isOnline = await getOnline();
async function createProject(cwd, options) {
	const { directory, template, language, linter, framework, packageManager, disableGit } = await p.group({
		directory: async () => {
			if (cwd) return normalizePosix(cwd);
			const defaultPath = "./";
			return await p.text({
				message: "Where would you like your project to be created?",
				placeholder: `  (hit Enter to use '${defaultPath}')`,
				defaultValue: defaultPath
			});
		},
		force: async ({ results: { directory } }) => {
			if (!options.dirCheck || !directory) return;
			if (!existsSync(directory)) return;
			const root = resolve(directory);
			const name = basename(root);
			const conflicts = readdirSync(root).filter((file) => !VALID_PROJECT_FILES.has(file) && !file.endsWith(".iml"));
			if (conflicts.length === 0) return;
			p.log.warn(`The directory ${pc.green(name)} contains files that could conflict:`);
			const conflictList = conflicts.map((file) => {
				try {
					return lstatSync(join(root, file)).isDirectory() ? ` ${pc.blue(file)}/` : ` ${file}`;
				} catch {
					return `${file}`;
				}
			}).join("\n");
			p.note(conflictList, "Potential Conflicts");
			const force = await p.confirm({
				message: `Directory is not empty. ${pc.yellow("Overwrite and continue?")}`,
				initialValue: false
			});
			if (p.isCancel(force) || !force) {
				p.cancel("Operation cancelled.");
				process.exit(0);
			}
		},
		template: async () => {
			if (options.template) return options.template;
			return await p.select({
				message: "Select which template you want to chose",
				options: templateOptions
			});
		},
		framework: async ({ results: { template } }) => {
			if (template === "custom-app") return "react";
			if (options.framework) return options.framework;
			return await p.select({
				message: "Select which framework you want to chose",
				options: frameworkOptions
			});
		},
		language: async () => {
			if (options.language) return options.language;
			return await p.select({
				message: "Select which language you want to chose",
				initialValue: "ts",
				options: languageOptions
			});
		},
		linter: async () => {
			if (options.linter) return options.linter;
			return await p.select({
				message: "Select which linter you want to chose",
				options: linterOptions
			});
		},
		packageManager: async () => {
			if (options.packageManager) return options.packageManager;
			return await p.select({
				message: "Select which package manager you want to choose to install packages",
				options: packageManagers.map((value) => ({
					value,
					title: value
				})),
				initialValue: getPackageManager()
			});
		},
		disableGit: async () => {
			if (!options.git) return options.git;
			return !await p.confirm({
				message: "Initialize a Git repository?",
				initialValue: true
			});
		}
	}, { onCancel: () => {
		p.cancel("Operation cancelled.");
		process.exit(0);
	} });
	const projectPath = resolve(directory);
	const projectName = await p.text({
		message: "Confirm or enter the package name:",
		initialValue: basename(projectPath).toLowerCase(),
		validate: validateProjectName
	});
	if (p.isCancel(projectName)) {
		p.cancel("Operation cancelled.");
		process.exit(0);
	}
	await create$1(projectPath, {
		name: projectName,
		template,
		language,
		linter,
		framework,
		packageManager,
		disableGit,
		install: options.install
	});
	p.log.success("Project created");
}
async function create$1(cwd, options) {
	try {
		mkdirp(cwd);
		setupTemplateFiles(options, cwd);
		await updateTypes(false, cwd);
		const pkgJSON = createPackageJSON(options);
		writePackageJSON(pkgJSON, cwd);
		chdir(cwd);
		await promptPackageInstallation(options, pkgJSON);
		if (!options.disableGit) if (tryGitInit(cwd)) p.log.success("Initialized a git repository");
		else p.log.error("Failed to initialize git repository");
		else p.log.info("Skipping git initialization");
	} catch (e) {
		const message = e instanceof Error ? e.message : "Unexpected Error";
		p.log.error(`Failed to scaffold ${options.template}: ${message}\n`);
		p.cancel("Exiting...");
		process.exit(1);
	}
}
async function promptPackageInstallation(options, pkg) {
	if (!options.install) return;
	const shouldInstall = await p.confirm({
		message: `Install dependencies using ${options.packageManager}?`,
		initialValue: true
	});
	if (p.isCancel(shouldInstall)) {
		p.cancel("Operation cancelled.");
		process.exit(0);
	}
	try {
		if (shouldInstall) {
			p.log.info(`Installing dependencies with ${options.packageManager}...`);
			await installPackages(options.packageManager, isOnline);
			const formatDeps = (deps) => {
				return Object.entries(deps).map(([name, version]) => `${pc.bold(name)}@${pc.blue(version)}`).join("\n");
			};
			if (pkg.dependencies) p.note(formatDeps(pkg.dependencies), "Dependencies");
			if (pkg.devDependencies) p.note(formatDeps(pkg.devDependencies), "Dev Dependencies");
			p.log.info("Dependencies installed successfully.");
		} else p.log.info(`Skipping install. You can run '${options.packageManager} install' later.`);
	} catch {
		p.log.info(`Failed to install dependencies, check the errors above.`);
	}
}

//#endregion
//#region src/commands/create.ts
const PathSchema = v.optional(v.string());
const OptionsSchema = v.strictObject({
	install: v.boolean(),
	git: v.boolean(),
	dirCheck: v.boolean(),
	language: v.optional(v.picklist(languageTypes)),
	template: v.optional(v.picklist(templateTypes)),
	framework: v.optional(v.picklist(frameworkTypes)),
	linter: v.optional(v.picklist(linterTypes)),
	packageManager: v.optional(v.picklist(packageManagers))
});
const templateOption = new Option("-t, --template <name>", "Template to use").choices(templateTypes);
const linterOption = new Option("--linter <name>", "Linter to use").choices(linterTypes);
const langOption = new Option("--language <lang>", "Language to use").choices(languageTypes);
const frameworkOption = new Option("--framework <name>", "Framework to use").choices(frameworkTypes);
const packageManagerOption = new Option("--pm, --package-manager <name>", "Package manager to use").choices(packageManagers);
const create = new Command("create").description("Scaffolds a new spicetify project").argument("[path]", "Where the project will be created").addOption(templateOption).addOption(langOption).addOption(frameworkOption).addOption(linterOption).addOption(packageManagerOption).option("--disable-git, --no-git", "Skip initializing a git repository.").option("--skip-install, --no-install", "Skip installing packages").option("--skip-dir-check, --no-dir-check", "Even if the folder is not empty, no prompt will be shown").action((path, opts) => {
	try {
		const cwd = safeParse(PathSchema, path);
		const options = safeParse(OptionsSchema, opts);
		runCommand(async () => {
			await createProject(cwd, options);
		}, {
			outro: true,
			message: (opts) => `Welcome to ${opts.name} CLI ${pc.gray(`(v${opts.version})`)}`
		});
	} catch (error) {
		if (v.isValiError(error)) {
			logger$2.error("\nInvalid configuration:");
			error.issues.forEach((issue) => {
				const path = issue.path?.map((p) => p.key).join(".") || "input";
				logger$2.error(`${pc.yellow(`  - ${path}:`)} ${issue.message}`);
			});
		} else logger$2.error(`\nAn unexpected error occurred: ${error instanceof Error ? error.message : String(error)}`);
		process.exit(1);
	}
});

//#endregion
//#region src/dev/server/templates/root.ts
const root = () => `<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Spicetify Creator | Dev Server</title>
</head>
<body>
  <h1>Spicetify Creator Server up and running.</h1>
</body>
</html>
`;

//#endregion
//#region src/dev/server/index.ts
function getWSPath(port) {
	return `/spicetify-creator-${port}`;
}
async function createHmrServer(config, logger = createLogger("hmrServer")) {
	const { port = DEFAULT_PORT, serveDir = outDir } = config;
	let isRunning = false;
	const mimeTypes = {
		".html": "text/html",
		".js": "text/javascript",
		".css": "text/css",
		".json": "application/json",
		".png": "image/png",
		".jpg": "image/jpeg",
		".jpeg": "image/jpeg",
		".gif": "image/gif",
		".svg": "image/svg+xml",
		".ico": "image/x-icon"
	};
	async function isPortAvailable(port) {
		return new Promise((resolve) => {
			const server = createServer$1();
			server.once("error", () => resolve(false));
			server.once("listening", () => {
				server.close(() => resolve(true));
			});
			server.listen(port);
		});
	}
	let actualPort = port;
	const portsToCheck = Array.from({ length: 10 }, (_, i) => port + i);
	const results = await Promise.all(portsToCheck.map(async (p) => ({
		port: p,
		available: await isPortAvailable(p)
	})));
	for (const result of results) {
		if (result.available) {
			actualPort = result.port;
			break;
		}
		logger.debug(`Port ${result.port} is in use, trying ${result.port + 1}...`);
	}
	if (actualPort !== port) logger.info(pc.yellow(`Port ${port} is in use, using port ${actualPort} instead`));
	const httpServer = createServer((req, res) => {
		const corsHeaders = {
			"Access-Control-Allow-Origin": "*",
			"Access-Control-Allow-Methods": "GET, POST, OPTIONS, PUT, PATCH, DELETE",
			"Access-Control-Allow-Headers": "X-Requested-With, Content-Type, Authorization",
			"Access-Control-Max-Age": "86400"
		};
		if (req.method === "OPTIONS") {
			res.writeHead(204, corsHeaders);
			res.end();
			return;
		}
		const wrapResponse = (statusCode, headers) => {
			res.writeHead(statusCode, {
				...headers,
				...corsHeaders
			});
		};
		const cleanUrl = (req.url ?? "/").split("?")[0] ?? "/";
		if (cleanUrl === "/" || cleanUrl === "/index.html") {
			wrapResponse(200, { "Content-Type": "text/html" });
			res.end(root());
			return;
		}
		const filePath = resolve(join(serveDir, cleanUrl.startsWith("/files/") ? cleanUrl.slice(6) : cleanUrl));
		try {
			if (existsSync(filePath) && statSync(filePath).isFile()) {
				const contentType = mimeTypes[extname(filePath).toLowerCase()] || "application/octet-stream";
				wrapResponse(200, { "Content-Type": contentType });
				createReadStream(filePath).pipe(res);
				return;
			}
		} catch {}
		wrapResponse(404, { "Content-Type": "text/html" });
		res.end("<h1>404 - Not Found</h1>");
	});
	const wss = new WebSocketServer({ noServer: true });
	const clients = /* @__PURE__ */ new Set();
	httpServer.on("upgrade", (req, socket, head) => {
		const { url } = req;
		if (!url?.startsWith(getWSPath(actualPort))) {
			socket.destroy();
			return;
		}
		wss.handleUpgrade(req, socket, head, (ws) => {
			clients.add(ws);
			ws.on("close", () => {
				clients.delete(ws);
			});
			ws.on("error", () => {
				clients.delete(ws);
			});
			wss.emit("connection", ws, req);
		});
	});
	function broadcast(data) {
		const message = JSON.stringify(data);
		for (const client of clients) if (client.readyState === WebSocket.OPEN) client.send(message);
	}
	return {
		start: async () => new Promise((resolve) => {
			httpServer.listen(actualPort, () => {
				logger.debug(`${pc.bold("HTTP Server Started at")}: ${pc.cyan(`http://localhost:${actualPort}/`)}`);
				isRunning = true;
				resolve();
			});
		}),
		stop: async () => new Promise((resolve, reject) => {
			if (!isRunning || !httpServer.listening) {
				isRunning = false;
				return resolve();
			}
			httpServer.close((err) => {
				isRunning = false;
				if (err) {
					if (err.code === "ERR_SERVER_NOT_RUNNING") return resolve();
					return reject(err);
				}
				logger.debug(`${pc.yellow("! ")} ${pc.gray("HTTP server stopped")}`);
				resolve();
			});
		}),
		broadcast,
		get port() {
			return actualPort;
		},
		get isRunning() {
			return isRunning;
		},
		get link() {
			return `http://localhost:${actualPort}`;
		},
		get wsLink() {
			return `ws://localhost:${actualPort}${getWSPath(actualPort)}`;
		}
	};
}

//#endregion
//#region src/utils/hmr.ts
const injectHMRExtension = async (rootLink, wsLink, outFiles, config, port) => {
	const extName = port === 54321 ? `sc-live-reload-helper.js` : `sc-live-reload-helper-${port}.js`;
	const spiceConfig = await getSpicetifyConfig();
	const cleanup = () => {
		logger$2.info(`Removing Live reload extension...`);
		try {
			runSpice([
				"config",
				"extensions",
				`${extName}-`
			]);
			runSpice(["apply"]);
			logger$2.info(pc.green(`${CHECK} Cleanup successful.`));
		} catch (e) {
			logger$2.error(pc.red(`${CROSS} Cleanup failed: `), e);
		}
		process.exit();
	};
	process.on("SIGINT", cleanup);
	process.on("SIGTERM", cleanup);
	try {
		logger$2.debug(`Preparing Live reload extension...`);
		const destDir = getExtensionDir();
		mkdirp(destDir);
		const outDir = resolve(destDir, extName);
		const { code } = await transform(readFileSync(liveReloadFilePath, "utf8"), {
			loader: "js",
			define: {
				_SERVER_URL: JSON.stringify(rootLink),
				_HOT_RELOAD_LINK: JSON.stringify(wsLink),
				_JS_PATH: JSON.stringify(`/files/${outFiles.js}`),
				_CSS_PATH: JSON.stringify(outFiles.css ? `/files/${outFiles.css}` : `/files/app.css`),
				_REMOVE_CMD: JSON.stringify(`spicetify config extensions ${extName}- && spicetify apply`),
				_CSS_ID: JSON.stringify(config.template === "extension" && config.cssId ? config.cssId : "sc-injected-css")
			}
		});
		writeFileSync(outDir, code);
		if (spiceConfig) {
			runSpice([
				"config",
				"extensions",
				extName
			]);
			runSpice(["apply"]);
		}
		logger$2.debug(pc.green(`${CHECK} Live reload extension injected successfully.`));
	} catch (err) {
		logger$2.error(pc.red(`${CROSS} Failed to inject HMR helper: ${err instanceof Error ? err.message : String(err)}`));
	}
};
const injectHMRCustomApp = async (rootLink, wsLink, outFiles, config, port) => {
	if (config.template !== "custom-app") throw new Error("only supports custom-app templates");
	const identifier = getEnName(config.name);
	const spiceConfig = await getSpicetifyConfig();
	const customAppId = port === 54321 ? urlSlugify(identifier) : `${urlSlugify(identifier)}-${port}`;
	const extName = port === 54321 ? `sc-live-reload-helper.js` : `sc-live-reload-helper-${port}.js`;
	const cleanup = () => {
		logger$2.info(`Removing Live reload custom-app...`);
		try {
			runSpice([
				"config",
				"extensions",
				`${extName}-`
			]);
			runSpice([
				"config",
				"custom_apps",
				`${customAppId}-`
			]);
			runSpice(["apply"]);
			logger$2.info(pc.green(`${CHECK} Cleanup successful.`));
		} catch (e) {
			logger$2.error(pc.red(`${CROSS} Cleanup failed: `), e);
		}
		process.exit();
	};
	process.on("SIGINT", cleanup);
	process.on("SIGTERM", cleanup);
	try {
		logger$2.debug(`Preparing Live reload custom-app...`);
		const destDir = resolve(getCustomAppsDir(), customAppId);
		mkdirp(destDir);
		const indexJsPath = resolve(destDir, "index.js");
		const extensionJsPath = resolve(destDir, outFiles.jsExtension ?? "extension.js");
		const { code: transformedIndexCode } = await transform(readFileSync(hmrCustomAppFilePath, "utf8"), {
			loader: "js",
			define: { _IMPORT_LINK: JSON.stringify(`${rootLink}/files/${outFiles.js}`) },
			platform: "browser"
		});
		writeFileSync(indexJsPath, transformedIndexCode);
		const { code: extensionCode } = await transform(readFileSync(liveReloadFilePath, "utf8"), {
			loader: "js",
			define: {
				_SERVER_URL: JSON.stringify(rootLink),
				_HOT_RELOAD_LINK: JSON.stringify(wsLink),
				_JS_PATH: JSON.stringify(`/files/${outFiles.jsExtension ?? "extension.js"}`),
				_CSS_PATH: JSON.stringify(outFiles.css ? `/files/${outFiles.css}` : `/files/app.css`),
				_REMOVE_CMD: JSON.stringify(`spicetify config custom_apps ${customAppId}- && spicetify apply`)
			}
		});
		writeFileSync(extensionJsPath, extensionCode);
		const icon = readFileSync(config.icon.default).toString();
		const activeIcon = config.icon.active ? readFileSync(config.icon.active).toString() : icon;
		const manifestPath = resolve(destDir, "manifest.json");
		const manifest = {
			name: config.name,
			subfiles: [],
			subfiles_extension: [outFiles.jsExtension ?? "extension.js"],
			icon,
			activeIcon
		};
		writeFileSync(manifestPath, JSON.stringify(manifest, null, 2));
		if (spiceConfig) {
			runSpice([
				"config",
				"custom_apps",
				customAppId
			]);
			runSpice(["apply"]);
		}
		logger$2.debug(pc.green(`${CHECK} Live reload custom-app injected successfully.`));
	} catch (err) {
		logger$2.error(pc.red(`${CROSS} Failed to inject HMR custom-app: ${err instanceof Error ? err.message : String(err)}`));
	}
};

//#endregion
//#region src/dev/index.ts
const DEFAULT_PORT = 54321;
const outDir = resolve(`./.spicetify/build/`);
async function dev$1(options) {
	logger$2.clear();
	logger$2.greeting(pc.green("Starting development environment"));
	let ctx;
	let server = void 0;
	await loadConfig(async (config, isNewUpdate) => {
		if (isNewUpdate) {
			logger$2.clear();
			logger$2.info(pc.green("Config updated, reloading..."));
			await server?.stop();
			server = void 0;
		}
		try {
			server = await createHmrServer({
				...config.serverConfig,
				serveDir: config.serverConfig.serveDir ?? outDir,
				port: options.port ?? config.serverConfig.port
			});
			await server.start();
			const outFiles = getOutFiles(config, true);
			if (config.template === "custom-app") await injectHMRCustomApp(server.link, server.wsLink, outFiles, config, server.port);
			else await injectHMRExtension(server.link, server.wsLink, outFiles, config, server.port);
			ctx = await context(getJSDevOptions(config, {
				...options,
				outFiles,
				server
			}));
			await ctx.watch();
			logger$2.info(pc.blue("Watching for changes..."));
		} catch (err) {
			logger$2.error("Failed to start dev server: ", err);
		}
		return async () => {
			await ctx?.dispose();
			ctx = void 0;
			await server?.stop();
		};
	});
}
function getJSDevOptions(config, options) {
	const entryPoints = getEntryPoints(config);
	const minify = false;
	const overrides = {
		...defaultBuildOptions,
		outdir: outDir,
		minify,
		sourcemap: "inline",
		external: [
			...config.esbuildOptions?.external ? config.esbuildOptions.external : [],
			"react",
			"react-dom"
		],
		define: {
			[DEV_MODE_VAR_NAME]: "true",
			...config.devModeVarName ? { [config.devModeVarName]: "true" } : {},
			...config.esbuildOptions.define
		},
		plugins: [...config.esbuildOptions?.plugins ? config.esbuildOptions.plugins : [], ...getCommonPlugins({
			...config,
			minify,
			buildOptions: {
				copy: true,
				apply: true,
				remove: config.template !== "custom-app",
				outDir
			},
			dev: true,
			server: options.server,
			outFiles: options.outFiles
		})]
	};
	return {
		entryPoints,
		...config.esbuildOptions,
		...overrides
	};
}

//#endregion
//#region src/commands/dev.ts
const CLIOptionsSchema = v.strictObject({ port: v.optional(v.pipe(v.union([v.string(), v.number()]), v.transform((val) => Number(val)), v.number("Port must be a valid number"), v.minValue(1, "Port must be greater than 0"), v.maxValue(65535, "Port must be less than 65536"))) });
const dev = new Command("dev").description("Develop your spicetify project").option("-p, --port <number>", "Port for the development server").action(async (opts) => {
	await dev$1(safeParse(CLIOptionsSchema, opts));
});

//#endregion
//#region src/commands/clean-spicetify.ts
function getHmrExtensions(port) {
	return port === 54321 ? "sc-live-reload-helper.js" : `sc-live-reload-helper-${port}.js`;
}
function getHmrCustomAppId(name, port) {
	const slugified = urlSlugify(getEnName(name));
	return port === 54321 ? slugified : `${slugified}-${port}`;
}
const cleanSpicetify = new Command("clean-spicetify").description("Remove HMR extensions and custom apps added by Spicetify Creator from Spicetify").option("-a, --all", "Clean all possible HMR artifacts").action(async (opts) => {
	const all = opts.all ?? false;
	try {
		const spiceConfig = await getSpicetifyConfig();
		const extensions = new Set(spiceConfig.AdditionalOptions.extensions);
		const customApps = new Set(spiceConfig.AdditionalOptions.custom_apps.split("|").filter(Boolean));
		const toRemove = [];
		if (all) for (let port = 54321; port <= 54330; port++) {
			const extName = getHmrExtensions(port);
			if (extensions.has(extName)) toRemove.push(extName);
		}
		else await (await loadConfig(async (config) => {
			const serverPort = config.serverConfig?.port ?? 54321;
			const extName = getHmrExtensions(serverPort);
			if (extensions.has(extName)) toRemove.push(extName);
			if (config.template === "custom-app") {
				const customAppId = getHmrCustomAppId(getEnName(config.name), serverPort);
				if (customApps.has(customAppId)) toRemove.push(customAppId);
			}
		})).unwatch();
		if (toRemove.length === 0) {
			console.log(pc.blue("No HMR artifacts found to clean."));
			return;
		}
		console.log(pc.yellow(`Removing: ${toRemove.join(", ")}...`));
		for (const item of toRemove) if (extensions.has(item)) runSpice([
			"config",
			"extensions",
			`${item}-`
		]);
		else runSpice([
			"config",
			"custom_apps",
			`${item}-`
		]);
		runSpice(["apply"]);
		console.log(pc.green(`${CHECK} Cleanup successful.`));
		process.exit(0);
	} catch (e) {
		console.error(pc.red(`${CROSS} Cleanup failed: `), e);
		process.exit(1);
	}
});

//#endregion
//#region src/commands/update-types.ts
const update_types = new Command("update-types").description("Update Spicetify Types").action(async () => {
	await updateTypes();
});

//#endregion
//#region src/bin.ts
logger$2.debug(`Env: ${JSON.stringify(env, null, 2)}\n`);
const command = new Command();
command.version(version).addCommand(create.alias("init")).addCommand(build$1).addCommand(dev).addCommand(cleanSpicetify.alias("clean-spice")).addCommand(update_types);
command.parse();

//#endregion
export {  };
