// @jsr/socali__modules/Signal.js
// Packages
import { FreeArray } from "./FreeArray.js";
// Classes
class Connection {
  // Private Properties
  ConnectionReferences;
  Location;
  Disconnected;
  // Constructor
  constructor(connections, callback){
    // Store our signal/callback
    this.ConnectionReferences = connections;
    // Store our initial disconnected state
    this.Disconnected = false;
    // Now store ourselves
    this.Location = connections.Push({
      Callback: callback,
      Connection: this
    });
  }
  // Public Methods
  Disconnect() {
    // Make sure we aren't already disconnected
    if (this.Disconnected) {
      return;
    }
    // Disconnect ourself
    this.Disconnected = true;
    // Remove ourselves from our signal
    this.ConnectionReferences.Remove(this.Location);
  }
  IsDisconnected() {
    return this.Disconnected;
  }
}
class Event {
  // Private Properties
  Signal;
  // Constructor
  constructor(signal){
    // Store our signal
    this.Signal = signal;
  }
  // Public Methods
  Connect(callback) {
    return this.Signal.Connect(callback);
  }
  IsDestroyed() {
    return this.Signal.IsDestroyed();
  }
}
class Signal {
  // Private Properties
  ConnectionReferences;
  DestroyedState;
  // Constructor
  constructor(){
    // Create our list of connections
    this.ConnectionReferences = new FreeArray();
    // Store our initial destroyed state
    this.DestroyedState = false;
  }
  // Public Methods
  Connect(callback) {
    // Make sure we aren't destroyed
    if (this.DestroyedState) {
      throw 'Cannot connect to a Destroyed Signal';
    }
    // Return our connection (since the connection handles everything itself)
    return new Connection(this.ConnectionReferences, callback);
  }
  Fire(...args) {
    // Make sure we aren't destroyed
    if (this.DestroyedState) {
      throw 'Cannot fire a Destroyed Signal';
    }
    // Loop through all of our connections
    for (const [_, reference] of this.ConnectionReferences.GetIterator()){
      // Fire our callback
      reference.Callback(...args);
    }
  }
  GetEvent() {
    // Return our event
    return new Event(this);
  }
  IsDestroyed() {
    return this.DestroyedState;
  }
  // Deconstructor
  Destroy() {
    // Make sure we aren't already destroyed
    if (this.DestroyedState) {
      return;
    }
    // Disconnect all of our connections (so that the connection is labeled as Disconnected)
    for (const [_, reference] of this.ConnectionReferences.GetIterator()){
      reference.Connection.Disconnect();
    }
    // Disconnect ourself
    this.DestroyedState = true;
  }
}
export const IsConnection = (value)=>{
  return value instanceof Connection;
};
export { Signal };
