// @jsr/3d__three/test/rollup.treeshake.config.js
import path from 'node:path';
import resolve from 'npm:@rollup/plugin-node-resolve@^15.0.1';
import filesize from 'npm:rollup-plugin-filesize@^10.0.0';
import terser from 'npm:@rollup/plugin-terser@^0.4.0';
import { visualizer } from 'npm:rollup-plugin-visualizer@^5.9.0';
import { glsl } from '../utils/build/rollup.config.js';
import chalk from 'npm:chalk@^5.2.0';

const statsFile = path.resolve( 'test/treeshake/stats.html' );

function logStatsFile() {

	return {
		writeBundle() {

			console.log();
			console.log( 'Open the following url in a browser to analyze the tree-shaken bundle.' );
			console.log( chalk.blue.bold.underline( statsFile ) );
			console.log();

		}
	};

}

export default [
	{
		input: 'test/treeshake/index.js',
		plugins: [
			resolve(),
		],
		output: [
			{
				format: 'esm',
				file: 'test/treeshake/index.bundle.js'
			}
		]
	},
	{
		input: 'test/treeshake/index.js',
		plugins: [
			resolve(),
			terser(),
			filesize( {
				showMinifiedSize: false,
			} ),
		],
		output: [
			{
				format: 'esm',
				file: 'test/treeshake/index.bundle.min.js'
			}
		]
	},
	{
		input: 'test/treeshake/index-src.js',
		plugins: [
			glsl(),
			terser(),
			visualizer( {
				filename: statsFile,
			} ),
			logStatsFile(),
		],
		output: [
			{
				format: 'esm',
				file: 'test/treeshake/index-src.bundle.min.js'
			}
		]
	},
];
